{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "669103b2-3d25-4d36-9cb3-d405323085dc",
   "metadata": {},
   "source": [
    "## contamination checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a522e26e-3137-4fe8-9ecb-fe98786c177e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,shutil,subprocess\n",
    "import glob\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 200)\n",
    "import pylab as plt\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from snipgenie import tools, aligners, app, trees, plotting, tools, rdiff\n",
    "from importlib import reload\n",
    "import gzip, tempfile\n",
    "import pylab as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f86c935-c9f9-4a89-a55a-32994f6f8b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.read_csv('/storage/btbgenie/monaghan/monaghan_results/samples.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a5aedc3-436d-4927-b037-805abf49e629",
   "metadata": {},
   "outputs": [],
   "source": [
    "#f1='/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/1579_S5_L001-4_R1_001.fastq.gz'\n",
    "#f2='/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/1579_S5_L001-4_R2_001.fastq.gz'\n",
    "f1='/storage/btbgenie/mbovis_ireland/contaminated/TB20-001410_S56_L001-4_R1_001.fastq.gz'\n",
    "f2='/storage/btbgenie/mbovis_ireland/contaminated/TB20-001410_S56_L001-4_R2_001.fastq.gz'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e5ee4968-81a5-42c4-9f1d-e0c498a9950c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def spades(file1, file2, path, outfile=None, threads=4):\n",
    "    \"\"\"Run spades\"\"\"\n",
    "    \n",
    "    cmd = 'spades -t %s --pe1-1 %s --pe1-2 %s --careful -o %s' %(threads,file1,file2,path)    \n",
    "    if not os.path.exists(path): \n",
    "        print (cmd)\n",
    "        subprocess.check_output(cmd, shell=True)\n",
    "    if outfile != None:\n",
    "        shutil.copy(os.path.join(path,'scaffolds.fasta'),outfile)\n",
    "    return outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e696d581-9c84-4347-999f-9c7fe15ee3e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#assemble \n",
    "name='1579'\n",
    "assembly_path='assembly'\n",
    "out = os.path.join(assembly_path,name+'.fa')\n",
    "spades(f1,f2, os.path.join(assembly_path,name), outfile=out, threads=12)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b145c44-ed04-40e4-840e-091ae1a4ed70",
   "metadata": {},
   "source": [
    "## kraken2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9995e823-fad2-424c-b859-d9506d2cd05b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kraken2 -db STANDARD16 --report krakenreport.txt --threads 4 --paired /storage/btbgenie/mbovis_ireland/contaminated/TB20-001410_S56_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/contaminated/TB20-001410_S56_L001-4_R2_001.fastq.gz > kraken.out\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading database information... done.\n",
      "5582279 sequences (1658.64 Mbp) processed in 27.070s (12373.0 Kseq/m, 3676.33 Mbp/m).\n",
      "  5433598 sequences classified (97.34%)\n",
      "  148681 sequences unclassified (2.66%)\n"
     ]
    }
   ],
   "source": [
    "def kraken(file1, file2, threads=4):\n",
    "    \"\"\"Run kraken2 on paired reads\"\"\"\n",
    "    \n",
    "    os.environ['KRAKEN2_DB_PATH'] = '/local/kraken2'\n",
    "    cmd = 'kraken2 -db STANDARD16 --report krakenreport.txt --threads {t} --paired {f1} {f2} > kraken.out'.format(f1=file1,f2=file2,t=threads)\n",
    "    print (cmd)\n",
    "    subprocess.check_output(cmd, shell=True)\n",
    "    rep=pd.read_csv('krakenreport.txt',sep='\\t',names=['perc_frag','n_frags_root','n_frags','rank_code','taxid','name'])   \n",
    "    rep['name'] = rep.name.str.lstrip()\n",
    "    return rep\n",
    "\n",
    "rep=kraken(f1, f2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bbd334fd-41bc-4b04-be78-fc150af878f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>perc_frag</th>\n",
       "      <th>n_frags_root</th>\n",
       "      <th>n_frags</th>\n",
       "      <th>rank_code</th>\n",
       "      <th>taxid</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>93.91</td>\n",
       "      <td>5242394</td>\n",
       "      <td>3950709</td>\n",
       "      <td>S</td>\n",
       "      <td>1764</td>\n",
       "      <td>Mycobacterium avium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    perc_frag  n_frags_root  n_frags rank_code  taxid                 name\n",
       "11      93.91       5242394  3950709         S   1764  Mycobacterium avium"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep=rep[rep.perc_frag>0.1]\n",
    "rep[rep.rank_code=='S']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea942f7-3f5c-409a-a6e4-ac85b51ada76",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "res=[]\n",
    "found = pd.read_csv('contam_monaghan_kraken.csv')\n",
    "samples = pd.read_csv('/storage/btbgenie/monaghan/monaghan_results/samples.csv')\n",
    "meta = pd.read_csv('/storage/btbgenie/monaghan/metadata/master.csv')\n",
    "for i,r in samples.iterrows():\n",
    "    name = r['sample']\n",
    "    #print (name)\n",
    "    if name in list(found['sample']):\n",
    "        continue\n",
    "    rep = kraken(r.filename1,r.filename2, 8)    \n",
    "    rep=rep[rep.perc_frag>0.1]    \n",
    "    rep['sample'] = name\n",
    "    res.append(rep)\n",
    "if len(res)>0:\n",
    "    df=pd.concat(res)\n",
    "    found = pd.concat([found,df])\n",
    "    found.to_csv('contam_monaghan_kraken.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aea1d3e2-8708-4ce4-8638-61c630590423",
   "metadata": {},
   "outputs": [],
   "source": [
    "#samples['meanGC'] = samples.apply( lambda x: tools.get_gc(x.filename1, limit=5e4).mean(), 1)\n",
    "samples.sample(150).meanGC.hist(bins=20,grid=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a76a0b-0283-4528-826f-c6fc86b51e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "x=found[found.rank_code=='G']\n",
    "contam = x.pivot_table(index='sample',columns='name',values='perc_frag').fillna(0)\n",
    "X=contam[contam.Mycobacterium<90]\n",
    "#X=X.merge(samples[['sample','meanGC']],left_index=True,right_on='sample')\n",
    "#X=X.set_index(['sample','meanGC'])\n",
    "\n",
    "ax=plt.subplots(1,1,figsize=(10,5))\n",
    "sns.heatmap(X,lw=1,cmap='Blues',annot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a9c091b-4d42-4e8e-b143-12fdb6375d09",
   "metadata": {},
   "outputs": [],
   "source": [
    "s1 = samples.sample(5)\n",
    "other=['2401','3825','040-Bov-2005-27.140']\n",
    "s = samples[samples['sample'].isin(clnames+other)]\n",
    "start=1473900\n",
    "end=1480000\n",
    "rdiff.show_coverage(s,'LT708304.1',start,end,app.mbovis_genome,app.mbovis_gb,minq=0,\n",
    "                      labelcol='sample',margin=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ba1b6376-0b4d-4462-96c3-e86d894cdce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hits</th>\n",
       "      <th>perc_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Mycobacterium bovis</th>\n",
       "      <td>3317</td>\n",
       "      <td>66.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paenibacillus polymyxa</th>\n",
       "      <td>14</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        hits  perc_hits\n",
       "Mycobacterium bovis     3317      66.34\n",
       "Paenibacillus polymyxa    14       0.28"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(app)\n",
    "reload(tools)\n",
    "app.fetch_contam_file()\n",
    "app.blast_contaminants(f1,limit=5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "141a86ae-d303-41d5-9fc7-cbb9c0eb368b",
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(tools)\n",
    "seqs = fastq_to_rec(f1,1000)\n",
    "SeqIO.write(seqs,'1579.fa','fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "76ab26ff-9735-495f-87c1-ac44b1f8897c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hits</th>\n",
       "      <th>perc_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Clostridium_tyrobutyricum</th>\n",
       "      <td>1266</td>\n",
       "      <td>63.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycobacterium_bovis</th>\n",
       "      <td>290</td>\n",
       "      <td>14.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paenibacillus_polymyxa</th>\n",
       "      <td>14</td>\n",
       "      <td>0.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clostridium_argentinense</th>\n",
       "      <td>10</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           hits  perc_hits\n",
       "Clostridium_tyrobutyricum  1266       63.3\n",
       "Mycobacterium_bovis         290       14.5\n",
       "Paenibacillus_polymyxa       14        0.7\n",
       "Clostridium_argentinense     10        0.5"
      ]
     },
     "execution_count": 366,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "app.blast_contaminants(f2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "2b6d1bc6-2ba2-4034-9d4b-9903d7de746f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clostridium_tyrobutyricum\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/home/farrell/.config/snipgenie/genome/removal.fa'"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def make_contam_targets(species=None, path=None):\n",
    "    \"\"\"Create targets from contaminant seqs for removal with bwa\"\"\"\n",
    "    \n",
    "    new = []    \n",
    "    seqs = SeqIO.parse('contam_genomes/contam.fa','fasta')\n",
    "    for seq in seqs:\n",
    "        title = seq.description.split('__')[0]\n",
    "        #print (title)\n",
    "        if species == None:\n",
    "            #don't include reference\n",
    "            if title != 'Mycobacterium_bovis':\n",
    "                new.append(seq)\n",
    "        elif title in species:\n",
    "            print (title)\n",
    "            new.append(seq)\n",
    "    if path==None:\n",
    "        path=tempfile.tempdir\n",
    "    filename = os.path.join(path,'removal.fa')\n",
    "    SeqIO.write(new,filename,'fasta')\n",
    "    return filename\n",
    "\n",
    "make_contam_targets(path=app.sequence_path,species=['Clostridium_tyrobutyricum'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "9c94b646-8ffd-469b-a605-fabc14496566",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this doesn't work!!\n",
    "reload(aligners)\n",
    "def remove_contaminants(f1, f2, outdir, species=None, threads=4):\n",
    "    \"\"\"Remove specified sequences from fastq files and keep unmapped\"\"\"\n",
    "    \n",
    "    targets = make_contam_targets(path=app.sequence_path, species=species)\n",
    "    aligners.build_bwa_index(targets)   \n",
    "    tmp = os.path.join(tempfile.tempdir,'contam_removal.bam')\n",
    "    idx = os.path.join(app.sequence_path,'removal.fa')\n",
    "    aligners.bwa_align(f1, f2, idx=idx, out=tmp, unmapped=outdir, threads=threads, overwrite=True)\n",
    "    os.remove(tmp)\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f235b7-6676-4ff0-a10a-07349d25b95d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "cleaned = '/storage/btbgenie/monaghan/cleaned2'\n",
    "remove_contaminants(f1, f2, cleaned, species=['Clostridium_tyrobutyricum'])#, 'Paenibacillus_polymyxa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "912e04d9-b5d2-40c9-adbd-0591a29b079e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10, 68440, 28, 15)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (147, 212, 402)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 912)\n",
      "[M::mem_pestat] mean and std.dev: (231.00, 113.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1167)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (218, 286, 357)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 635)\n",
      "[M::mem_pestat] mean and std.dev: (287.31, 107.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 774)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (304, 1434, 2491)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6865)\n",
      "[M::mem_pestat] mean and std.dev: (1317.12, 1130.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9052)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (179, 263, 465)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1037)\n",
      "[M::mem_pestat] mean and std.dev: (367.27, 252.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1378)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538374 reads in 45.753 CPU sec, 5.612 real sec\n",
      "[M::process] read 538264 sequences (80000264 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 69404, 16, 14)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (223, 311, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9724)\n",
      "[M::mem_pestat] mean and std.dev: (1183.93, 1580.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12891)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (215, 283, 350)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 620)\n",
      "[M::mem_pestat] mean and std.dev: (282.42, 105.93)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 755)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (834, 1434, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6231)\n",
      "[M::mem_pestat] mean and std.dev: (1687.19, 1029.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8030)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (117, 222, 288)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 630)\n",
      "[M::mem_pestat] mean and std.dev: (197.54, 89.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 801)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 539258 reads in 45.477 CPU sec, 5.585 real sec\n",
      "[M::process] read 538936 sequences (80000074 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (21, 68386, 26, 16)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 432, 793)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1921)\n",
      "[M::mem_pestat] mean and std.dev: (415.28, 363.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2485)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (219, 288, 358)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 636)\n",
      "[M::mem_pestat] mean and std.dev: (289.24, 108.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 775)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2395, 4421)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10395)\n",
      "[M::mem_pestat] mean and std.dev: (3202.50, 2886.70)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14749)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (162, 230, 322)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 642)\n",
      "[M::mem_pestat] mean and std.dev: (246.80, 126.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 802)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538264 reads in 45.517 CPU sec, 5.582 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (16, 69367, 17, 9)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (205, 311, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9760)\n",
      "[M::mem_pestat] mean and std.dev: (1042.06, 1359.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12945)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (217, 285, 355)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 631)\n",
      "[M::mem_pestat] mean and std.dev: (286.44, 107.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 769)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (549, 1434, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6726)\n",
      "[M::mem_pestat] mean and std.dev: (1451.94, 988.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8785)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 538936 reads in 44.697 CPU sec, 5.591 real sec\n",
      "[M::process] read 538776 sequences (80000203 bp)...\n",
      "[M::process] read 417450 sequences (61887137 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 69535, 14, 18)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (203, 238, 349)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 641)\n",
      "[M::mem_pestat] mean and std.dev: (244.85, 76.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 787)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (218, 287, 356)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 632)\n",
      "[M::mem_pestat] mean and std.dev: (287.59, 108.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 770)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (2072, 2418, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (950, 3755)\n",
      "[M::mem_pestat] mean and std.dev: (2350.40, 344.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (389, 4316)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (206, 302, 359)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 665)\n",
      "[M::mem_pestat] mean and std.dev: (257.12, 114.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 818)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538776 reads in 44.815 CPU sec, 5.523 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 54365, 12, 11)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (270, 307, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9630)\n",
      "[M::mem_pestat] mean and std.dev: (1493.13, 1650.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12750)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (216, 284, 353)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 627)\n",
      "[M::mem_pestat] mean and std.dev: (284.26, 106.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 764)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1572.40, 954.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (276, 306, 611)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1281)\n",
      "[M::mem_pestat] mean and std.dev: (390.09, 188.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1616)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 417450 reads in 34.441 CPU sec, 4.294 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2061_S59_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2061_S59_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 46.721 sec; CPU: 352.410 sec\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/2061_S59_L001-4_R1_001.fastq.gz -2 cleaned/2061_S59_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 2506362 reads\n",
      "[bwa_index] Pack FASTA... 0.15 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.39 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.11 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.32 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 19.191 sec; CPU: 19.114 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2423p_S29_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2423p_S29_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 546676 sequences (80000281 bp)...\n",
      "[M::process] read 548190 sequences (80000029 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 16200, 19, 6)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (87, 157, 229)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 513)\n",
      "[M::mem_pestat] mean and std.dev: (164.71, 93.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1438, 2395, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4948)\n",
      "[M::mem_pestat] mean and std.dev: (1995.50, 762.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6118)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 546676 reads in 45.540 CPU sec, 5.617 real sec\n",
      "[M::process] read 548892 sequences (80000189 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 16086, 17, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (83, 152, 223)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 503)\n",
      "[M::mem_pestat] mean and std.dev: (160.38, 92.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 643)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1830.65, 906.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 548190 reads in 46.445 CPU sec, 5.700 real sec\n",
      "[M::process] read 545930 sequences (80000232 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 16265, 17, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (85, 151, 222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 496)\n",
      "[M::mem_pestat] mean and std.dev: (160.45, 91.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 633)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1416, 2307, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4992)\n",
      "[M::mem_pestat] mean and std.dev: (1746.38, 785.70)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6184)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 548892 reads in 45.941 CPU sec, 5.635 real sec\n",
      "[M::process] read 548538 sequences (80000130 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 16108, 14, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (88, 158, 231)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 517)\n",
      "[M::mem_pestat] mean and std.dev: (166.33, 94.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 660)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (549, 2394, 2419)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6159)\n",
      "[M::mem_pestat] mean and std.dev: (1500.92, 912.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8029)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 545930 reads in 46.072 CPU sec, 5.652 real sec\n",
      "[M::process] read 547466 sequences (80000225 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 16599, 24, 6)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (85, 151, 224)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 502)\n",
      "[M::mem_pestat] mean and std.dev: (160.97, 91.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 641)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (549, 2491, 2645)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6837)\n",
      "[M::mem_pestat] mean and std.dev: (1542.16, 944.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8933)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 548538 reads in 45.225 CPU sec, 5.565 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 16239, 18, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (84, 152, 226)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 510)\n",
      "[M::mem_pestat] mean and std.dev: (161.83, 93.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 652)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (549, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6801)\n",
      "[M::mem_pestat] mean and std.dev: (1913.67, 1501.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8885)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 547466 reads in 44.762 CPU sec, 5.603 real sec\n",
      "[M::process] read 546470 sequences (80000248 bp)...\n",
      "[M::process] read 547304 sequences (80000228 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 16333, 15, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (87, 157, 229)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 513)\n",
      "[M::mem_pestat] mean and std.dev: (165.33, 94.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (635, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6629)\n",
      "[M::mem_pestat] mean and std.dev: (1715.40, 985.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8627)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 546470 reads in 45.147 CPU sec, 5.549 real sec\n",
      "[M::process] read 547262 sequences (80000102 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 16249, 20, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (87, 155, 226)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 504)\n",
      "[M::mem_pestat] mean and std.dev: (163.58, 93.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 643)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1418, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5063)\n",
      "[M::mem_pestat] mean and std.dev: (1623.82, 818.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6278)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 547304 reads in 45.658 CPU sec, 5.608 real sec\n",
      "[M::process] read 548764 sequences (80000297 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 16133, 13, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (86, 156, 228)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 512)\n",
      "[M::mem_pestat] mean and std.dev: (164.71, 94.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 654)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2608, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1957.55, 840.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 547262 reads in 45.665 CPU sec, 5.584 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 16268, 17, 6)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (84, 152, 223)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 501)\n",
      "[M::mem_pestat] mean and std.dev: (160.99, 92.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 640)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (549, 1434, 2638)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6816)\n",
      "[M::mem_pestat] mean and std.dev: (1599.47, 963.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8905)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 548764 reads in 44.861 CPU sec, 5.614 real sec\n",
      "[M::process] read 212540 sequences (30959572 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 6212, 8, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (83, 152, 221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
      "[M::mem_pestat] mean and std.dev: (159.45, 91.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 635)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 212540 reads in 17.740 CPU sec, 2.205 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2423p_S29_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/2423p_S29_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 61.290 sec; CPU: 475.225 sec\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/2423p_S29_L001-4_R1_001.fastq.gz -2 cleaned/2423p_S29_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 4422614 reads\n",
      "[bwa_index] Pack FASTA... 0.15 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.33 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.08 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 18.880 sec; CPU: 18.800 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3496_S83_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3496_S83_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 538630 sequences (80000134 bp)...\n",
      "[M::process] read 538174 sequences (80000257 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (22, 41489, 19, 11)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (208, 430, 1621)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4447)\n",
      "[M::mem_pestat] mean and std.dev: (1223.05, 1276.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6327)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (179, 252, 327)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 623)\n",
      "[M::mem_pestat] mean and std.dev: (254.04, 112.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 771)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (561, 1434, 2395)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6063)\n",
      "[M::mem_pestat] mean and std.dev: (1431.67, 1047.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7897)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (116, 231, 556)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1436)\n",
      "[M::mem_pestat] mean and std.dev: (211.44, 140.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1876)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538630 reads in 45.360 CPU sec, 5.611 real sec\n",
      "[M::process] read 195252 sequences (28988248 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (21, 41523, 15, 14)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (201, 247, 685)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1653)\n",
      "[M::mem_pestat] mean and std.dev: (317.16, 196.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2137)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (181, 253, 329)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 625)\n",
      "[M::mem_pestat] mean and std.dev: (255.85, 112.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 773)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (561, 803, 2493)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6357)\n",
      "[M::mem_pestat] mean and std.dev: (958.25, 737.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8289)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (267, 363, 561)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1149)\n",
      "[M::mem_pestat] mean and std.dev: (392.33, 250.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1443)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538174 reads in 46.477 CPU sec, 5.782 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 15153, 5, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (179, 252, 326)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 620)\n",
      "[M::mem_pestat] mean and std.dev: (254.53, 113.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 767)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 195252 reads in 17.613 CPU sec, 2.178 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3496_S83_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3496_S83_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 15.004 sec; CPU: 110.295 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/3496_S83_L001-4_R1_001.fastq.gz -2 cleaned/3496_S83_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 849242 reads\n",
      "[bwa_index] Pack FASTA... 0.15 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.32 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.07 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 18.874 sec; CPU: 18.786 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/362p_S18_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/362p_S18_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 535894 sequences (80000004 bp)...\n",
      "[M::process] read 536458 sequences (80000177 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5, 11620, 21, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (132, 224, 319)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 693)\n",
      "[M::mem_pestat] mean and std.dev: (231.21, 128.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 880)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (798, 2398, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6303)\n",
      "[M::mem_pestat] mean and std.dev: (1710.42, 883.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8138)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535894 reads in 49.281 CPU sec, 6.078 real sec\n",
      "[M::process] read 537238 sequences (80000181 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 11499, 13, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (124, 217, 307)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 673)\n",
      "[M::mem_pestat] mean and std.dev: (222.10, 124.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 856)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2408, 6837)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 17643)\n",
      "[M::mem_pestat] mean and std.dev: (3369.31, 3113.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 23046)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536458 reads in 49.934 CPU sec, 6.128 real sec\n",
      "[M::process] read 537078 sequences (80000104 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 11528, 18, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (121, 213, 303)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 667)\n",
      "[M::mem_pestat] mean and std.dev: (218.93, 123.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 849)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (798, 2221, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6228)\n",
      "[M::mem_pestat] mean and std.dev: (1565.13, 834.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8038)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 537238 reads in 49.728 CPU sec, 6.105 real sec\n",
      "[M::process] read 535746 sequences (80000106 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 11437, 10, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (122, 214, 304)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 668)\n",
      "[M::mem_pestat] mean and std.dev: (219.12, 123.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 850)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1838.12, 656.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 537078 reads in 49.532 CPU sec, 6.078 real sec\n",
      "[M::process] read 537102 sequences (80000274 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5, 11678, 16, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (131, 227, 316)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 686)\n",
      "[M::mem_pestat] mean and std.dev: (230.84, 127.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 871)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (798, 2395, 2687)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6465)\n",
      "[M::mem_pestat] mean and std.dev: (1523.23, 881.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8354)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535746 reads in 49.874 CPU sec, 6.122 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 11785, 19, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (122, 214, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 671)\n",
      "[M::mem_pestat] mean and std.dev: (220.51, 124.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 854)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (2280, 2633, 2687)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1466, 3501)\n",
      "[M::mem_pestat] mean and std.dev: (2567.20, 125.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1059, 3908)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 537102 reads in 51.274 CPU sec, 6.343 real sec\n",
      "[M::process] read 536496 sequences (80000127 bp)...\n",
      "[M::process] read 536130 sequences (80000094 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 11753, 16, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (125, 218, 309)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 677)\n",
      "[M::mem_pestat] mean and std.dev: (223.29, 124.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 861)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (812, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6275)\n",
      "[M::mem_pestat] mean and std.dev: (1725.57, 967.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8096)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536496 reads in 50.236 CPU sec, 6.172 real sec\n",
      "[M::process] read 536324 sequences (80000177 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 11589, 14, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (130, 224, 315)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 685)\n",
      "[M::mem_pestat] mean and std.dev: (229.01, 127.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 870)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (531, 2394, 2491)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6411)\n",
      "[M::mem_pestat] mean and std.dev: (1635.77, 916.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8371)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536130 reads in 50.231 CPU sec, 6.175 real sec\n",
      "[M::process] read 536446 sequences (80000253 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 11645, 18, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (124, 217, 308)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 676)\n",
      "[M::mem_pestat] mean and std.dev: (223.57, 126.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 860)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (667, 2491, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6565)\n",
      "[M::mem_pestat] mean and std.dev: (1860.24, 1141.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8531)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536324 reads in 50.608 CPU sec, 6.212 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 11527, 12, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (125, 219, 310)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 680)\n",
      "[M::mem_pestat] mean and std.dev: (225.50, 126.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 865)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2077, 2408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4356)\n",
      "[M::mem_pestat] mean and std.dev: (1767.00, 624.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5330)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536446 reads in 49.857 CPU sec, 6.178 real sec\n",
      "[M::process] read 535736 sequences (80000136 bp)...\n",
      "[M::process] read 536978 sequences (80000238 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 11543, 19, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (128, 227, 319)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 701)\n",
      "[M::mem_pestat] mean and std.dev: (231.04, 129.87)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 892)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1415, 1435, 2629)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5057)\n",
      "[M::mem_pestat] mean and std.dev: (1523.00, 836.74)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6271)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535736 reads in 50.146 CPU sec, 6.160 real sec\n",
      "[M::process] read 537106 sequences (80000268 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 11613, 15, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (124, 216, 305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 667)\n",
      "[M::mem_pestat] mean and std.dev: (222.50, 124.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 848)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (495, 1434, 2491)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6483)\n",
      "[M::mem_pestat] mean and std.dev: (1499.29, 891.87)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8479)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 536978 reads in 50.160 CPU sec, 6.154 real sec\n",
      "[M::process] read 217084 sequences (32312193 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 11463, 13, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (119, 213, 303)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 671)\n",
      "[M::mem_pestat] mean and std.dev: (219.17, 125.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 855)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2491, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1912.00, 790.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 537106 reads in 50.330 CPU sec, 6.243 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 4644, 4, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (114, 212, 301)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 675)\n",
      "[M::mem_pestat] mean and std.dev: (216.49, 125.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 862)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 217084 reads in 19.945 CPU sec, 2.497 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/362p_S18_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/362p_S18_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 88.162 sec; CPU: 673.070 sec\n",
      "[bam_sort_core] merging from 3 files and 1 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/362p_S18_L001-4_R1_001.fastq.gz -2 cleaned/362p_S18_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 5767352 reads\n",
      "[bwa_index] Pack FASTA... 0.16 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.43 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.11 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 19.022 sec; CPU: 18.950 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3825_S62_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3825_S62_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 545078 sequences (80000206 bp)...\n",
      "[M::process] read 479970 sequences (70493492 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 9367, 17, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 275, 371)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 763)\n",
      "[M::mem_pestat] mean and std.dev: (278.16, 143.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 959)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2218, 2687)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5193)\n",
      "[M::mem_pestat] mean and std.dev: (1617.86, 855.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6446)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 545078 reads in 44.706 CPU sec, 5.534 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7, 8054, 10, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (175, 276, 374)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 772)\n",
      "[M::mem_pestat] mean and std.dev: (279.41, 145.41)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 971)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (2491, 2608, 6770)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 15328)\n",
      "[M::mem_pestat] mean and std.dev: (4200.20, 3142.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 19607)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 479970 reads in 39.619 CPU sec, 4.948 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3825_S62_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/3825_S62_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 12.766 sec; CPU: 85.311 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/3825_S62_L001-4_R1_001.fastq.gz -2 cleaned/3825_S62_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 863590 reads\n",
      "[bwa_index] Pack FASTA... 0.15 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.28 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.11 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.13 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 18.906 sec; CPU: 18.821 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4798_S58_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4798_S58_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 541988 sequences (80000268 bp)...\n",
      "[M::process] read 541914 sequences (80000193 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 58932, 20, 13)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (155, 218, 281)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 533)\n",
      "[M::mem_pestat] mean and std.dev: (220.92, 96.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (435, 860, 2395)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6315)\n",
      "[M::mem_pestat] mean and std.dev: (1029.79, 869.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8275)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (195, 373, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9780)\n",
      "[M::mem_pestat] mean and std.dev: (1074.31, 1480.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12975)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 541988 reads in 44.288 CPU sec, 5.478 real sec\n",
      "[M::process] read 339758 sequences (50183946 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (14, 58110, 28, 13)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (125, 159, 269)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 557)\n",
      "[M::mem_pestat] mean and std.dev: (185.00, 89.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 701)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (156, 217, 281)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 531)\n",
      "[M::mem_pestat] mean and std.dev: (220.60, 95.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 656)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (364, 651, 2490)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6742)\n",
      "[M::mem_pestat] mean and std.dev: (1186.93, 1148.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8868)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (156, 236, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9858)\n",
      "[M::mem_pestat] mean and std.dev: (1098.92, 1356.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13092)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 541914 reads in 44.377 CPU sec, 5.495 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7, 36610, 18, 8)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (158, 218, 282)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 530)\n",
      "[M::mem_pestat] mean and std.dev: (221.84, 95.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 654)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 874, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7239)\n",
      "[M::mem_pestat] mean and std.dev: (2062.44, 2301.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11270)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 339758 reads in 27.974 CPU sec, 3.490 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4798_S58_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4798_S58_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 16.329 sec; CPU: 117.561 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/4798_S58_L001-4_R1_001.fastq.gz -2 cleaned/4798_S58_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 874676 reads\n",
      "[bwa_index] Pack FASTA... 0.14 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.19 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.23 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 18.889 sec; CPU: 18.817 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4948_S8_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4948_S8_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 541928 sequences (80000071 bp)...\n",
      "[M::process] read 64754 sequences (9543186 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 69580, 21, 15)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (292, 345, 751)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1669)\n",
      "[M::mem_pestat] mean and std.dev: (444.00, 332.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2128)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 362, 449)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 861)\n",
      "[M::mem_pestat] mean and std.dev: (345.15, 146.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1067)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (534, 1091, 4547)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 12573)\n",
      "[M::mem_pestat] mean and std.dev: (2553.62, 2763.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 16586)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 404, 512)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1056)\n",
      "[M::mem_pestat] mean and std.dev: (323.17, 132.87)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1328)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 541928 reads in 44.707 CPU sec, 5.618 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 8463, 4, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 362, 448)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 856)\n",
      "[M::mem_pestat] mean and std.dev: (343.57, 146.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1060)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 64754 reads in 5.513 CPU sec, 0.683 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4948_S8_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4948_S8_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 8.462 sec; CPU: 51.189 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/4948_S8_L001-4_R1_001.fastq.gz -2 cleaned/4948_S8_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 358154 reads\n",
      "[bwa_index] Pack FASTA... 0.15 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.16 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.14 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 5.96 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 18.589 sec; CPU: 18.522 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4960_S92_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4960_S92_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 553968 sequences (80000190 bp)...\n",
      "[M::process] read 552496 sequences (80000136 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 15464, 16, 8)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (92, 181, 275)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 641)\n",
      "[M::mem_pestat] mean and std.dev: (192.97, 119.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 824)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2394, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4956)\n",
      "[M::mem_pestat] mean and std.dev: (1807.33, 805.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6130)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 553968 reads in 57.516 CPU sec, 7.142 real sec\n",
      "[M::process] read 182638 sequences (26356016 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10, 15347, 19, 8)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (290, 362, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9590)\n",
      "[M::mem_pestat] mean and std.dev: (1362.70, 1649.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12690)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (92, 183, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 647)\n",
      "[M::mem_pestat] mean and std.dev: (194.20, 120.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 832)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (886, 1434, 2491)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5701)\n",
      "[M::mem_pestat] mean and std.dev: (1394.41, 745.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7306)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 552496 reads in 59.590 CPU sec, 7.430 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4, 5086, 4, 4)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (88, 180, 273)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 643)\n",
      "[M::mem_pestat] mean and std.dev: (191.41, 118.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 828)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 182638 reads in 19.208 CPU sec, 2.414 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4960_S92_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4960_S92_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 18.367 sec; CPU: 137.184 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/4960_S92_L001-4_R1_001.fastq.gz -2 cleaned/4960_S92_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 970512 reads\n",
      "[bwa_index] Pack FASTA... 0.16 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 14.05 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.11 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.49 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 21.024 sec; CPU: 20.956 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4973_S49_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4973_S49_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 538228 sequences (80000104 bp)...\n",
      "[M::process] read 537886 sequences (80000057 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 12439, 14, 11)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (115, 214, 314)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 712)\n",
      "[M::mem_pestat] mean and std.dev: (224.53, 135.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 911)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (531, 1434, 2395)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6123)\n",
      "[M::mem_pestat] mean and std.dev: (1368.92, 903.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7987)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (326, 469, 1097)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2639)\n",
      "[M::mem_pestat] mean and std.dev: (426.44, 275.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3410)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 538228 reads in 50.388 CPU sec, 6.239 real sec\n",
      "[M::process] read 73992 sequences (10988363 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (11, 12243, 16, 8)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (717, 793, 1129)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1953)\n",
      "[M::mem_pestat] mean and std.dev: (737.00, 487.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2686)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (118, 221, 322)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 730)\n",
      "[M::mem_pestat] mean and std.dev: (229.46, 137.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 934)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (667, 2633, 6759)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 18943)\n",
      "[M::mem_pestat] mean and std.dev: (3640.56, 3138.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 25035)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 537886 reads in 59.955 CPU sec, 7.481 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 1721, 1, 1)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (117, 218, 317)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 717)\n",
      "[M::mem_pestat] mean and std.dev: (227.00, 133.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 917)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 73992 reads in 8.955 CPU sec, 1.110 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4973_S49_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_24-09-19/4973_S49_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 16.975 sec; CPU: 120.315 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/4973_S49_L001-4_R1_001.fastq.gz -2 cleaned/4973_S49_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 908390 reads\n",
      "[bwa_index] Pack FASTA... 0.16 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=87866752, availableWord=18182240\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 29991792 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 55405792 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 77989824 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 35 iterations.\n",
      "[bwa_index] 12.78 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.15 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.10 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 6.51 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "[main] Real time: 19.779 sec; CPU: 19.707 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bwa index /home/farrell/.config/snipgenie/genome/removal.fa\n",
      "bwa mem -M -t 8  /home/farrell/.config/snipgenie/genome/removal.fa \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/548_S25_L001-4_R1_001.fastq.gz\" \"/storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/548_S25_L001-4_R2_001.fastq.gz\" | samtools view  -bt - | samtools sort -o /tmp/contam_removal.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::process] read 535170 sequences (80000228 bp)...\n",
      "[M::process] read 535712 sequences (80000069 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5, 13531, 21, 6)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (146, 242, 327)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 689)\n",
      "[M::mem_pestat] mean and std.dev: (242.46, 126.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 870)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2233, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5031)\n",
      "[M::mem_pestat] mean and std.dev: (1776.58, 825.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6230)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535170 reads in 48.965 CPU sec, 6.042 real sec\n",
      "[M::process] read 535380 sequences (80000216 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 13573, 26, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (140, 233, 319)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 677)\n",
      "[M::mem_pestat] mean and std.dev: (234.33, 124.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 856)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (632, 1434, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6560)\n",
      "[M::mem_pestat] mean and std.dev: (1326.18, 858.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8536)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535712 reads in 49.720 CPU sec, 6.105 real sec\n",
      "[M::process] read 535454 sequences (80000219 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 13378, 27, 6)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (145, 239, 324)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 682)\n",
      "[M::mem_pestat] mean and std.dev: (238.94, 124.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 861)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (912, 2395, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6075)\n",
      "[M::mem_pestat] mean and std.dev: (1698.30, 1060.98)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7796)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535380 reads in 49.572 CPU sec, 6.087 real sec\n",
      "[M::process] read 534934 sequences (80000168 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5, 13575, 17, 10)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (142, 238, 322)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 682)\n",
      "[M::mem_pestat] mean and std.dev: (237.78, 124.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 862)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2509, 6802)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 17538)\n",
      "[M::mem_pestat] mean and std.dev: (4112.76, 3255.85)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 22906)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (218, 3390, 3390)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9734)\n",
      "[M::mem_pestat] mean and std.dev: (1957.00, 1756.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12906)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 535454 reads in 49.631 CPU sec, 6.095 real sec\n",
      "[M::process] read 535348 sequences (80000096 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 13434, 25, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (149, 244, 328)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 686)\n",
      "[M::mem_pestat] mean and std.dev: (242.76, 124.80)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 865)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2402, 2638)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5046)\n",
      "[M::mem_pestat] mean and std.dev: (1848.59, 807.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6250)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 534934 reads in 49.855 CPU sec, 6.126 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 13476, 18, 3)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (143, 236, 322)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 680)\n",
      "[M::mem_pestat] mean and std.dev: (236.23, 123.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 859)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (928, 2633, 6620)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 18004)\n",
      "[M::mem_pestat] mean and std.dev: (3409.33, 2995.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 23696)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535348 reads in 49.698 CPU sec, 6.166 real sec\n",
      "[M::process] read 534902 sequences (80000053 bp)...\n",
      "[M::process] read 535748 sequences (80000108 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 13337, 22, 7)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (149, 241, 327)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 683)\n",
      "[M::mem_pestat] mean and std.dev: (242.26, 125.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 861)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (803, 2283, 2633)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6293)\n",
      "[M::mem_pestat] mean and std.dev: (1616.78, 809.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8123)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 534902 reads in 49.690 CPU sec, 6.094 real sec\n",
      "[M::process] read 140368 sequences (20943837 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 13629, 19, 5)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (140, 233, 317)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 671)\n",
      "[M::mem_pestat] mean and std.dev: (234.24, 123.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 848)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1434, 2303, 2608)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4956)\n",
      "[M::mem_pestat] mean and std.dev: (1793.12, 719.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6130)\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 535748 reads in 49.057 CPU sec, 6.092 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 3434, 8, 2)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (136, 231, 314)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 670)\n",
      "[M::mem_pestat] mean and std.dev: (232.14, 123.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 848)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 140368 reads in 13.149 CPU sec, 1.626 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -M -t 8 /home/farrell/.config/snipgenie/genome/removal.fa /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/548_S25_L001-4_R1_001.fastq.gz /storage/btbgenie/mbovis_ireland/Monaghan/Fastqs_29-07-19/548_S25_L001-4_R2_001.fastq.gz\n",
      "[main] Real time: 56.460 sec; CPU: 410.875 sec\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b -f12 /tmp/contam_removal.bam | samtools fastq -1 cleaned/548_S25_L001-4_R1_001.fastq.gz -2 cleaned/548_S25_L001-4_R2_001.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::bam2fq_mainloop] discarded 0 singletons\n",
      "[M::bam2fq_mainloop] processed 3433264 reads\n"
     ]
    }
   ],
   "source": [
    "clnames=list(X.index)\n",
    "csamples = samples[samples['sample'].isin(cleaned)]\n",
    "for i,r in csamples.iterrows():\n",
    "    remove_contaminants(r.filename1, r.filename2, cleaned, threads=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdad1300-2f45-4127-829a-4335c801aa32",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "args = {'threads':12, 'outdir': cleaned+'/snipgenie_results', 'labelsep':'_',\n",
    "        'input':[cleaned],\n",
    "         'species': 'Mbovis-AF212297',\n",
    "         'custom_filters': True,\n",
    "         'buildtree':True}\n",
    "W = app.WorkFlow(**args)\n",
    "st = W.setup()\n",
    "W.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "85ad482a-063f-4611-ba31-42a5ac40995a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"td020349b37c44d9f8995aadd80b31274\" style=\"text-align:center\"><svg class=\"toyplot-canvas-Canvas\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns=\"http://www.w3.org/2000/svg\" width=\"1000.0px\" height=\"500.0px\" viewBox=\"0 0 1000.0 500.0\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" id=\"t101fcba8df274600abf79d1f60a3f9de\"><g class=\"toyplot-coordinates-Cartesian\" id=\"tb1d08f21353e4b26a63ec6869a7d96e9\"><clipPath id=\"t91d97ef5e21b443f8434a52213468e26\"><rect x=\"30.0\" y=\"30.0\" width=\"940.0\" height=\"440.0\"></rect></clipPath><g clip-path=\"url(#t91d97ef5e21b443f8434a52213468e26)\"><g class=\"toytree-mark-Toytree\" id=\"t26b8189d87764883af638bf62524708e\"><g class=\"toytree-Edges\" style=\"fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2\"><path d=\"M 51.9 116.5 L 51.9 168.5 L 60.1 168.5\" id=\"30,29\"></path><path d=\"M 60.1 168.5 L 60.1 110.9 L 513.5 110.9\" id=\"29,28\"></path><path d=\"M 60.1 168.5 L 60.1 226.2 L 88.7 226.2\" id=\"29,27\"></path><path d=\"M 513.5 110.9 L 513.5 132.5 L 526.8 132.5\" id=\"28,26\"></path><path d=\"M 88.7 226.2 L 88.7 264.3 L 88.7 264.3\" id=\"27,25\"></path><path d=\"M 526.8 132.5 L 526.8 151.1 L 593.6 151.1\" id=\"26,24\"></path><path d=\"M 88.7 264.3 L 88.7 225.3 L 847.0 225.3\" id=\"25,23\"></path><path d=\"M 88.7 264.3 L 88.7 303.3 L 731.8 303.3\" id=\"25,22\"></path><path d=\"M 731.8 303.3 L 731.8 344.3 L 731.9 344.3\" id=\"22,21\"></path><path d=\"M 731.9 344.3 L 731.9 305.7 L 740.7 305.7\" id=\"21,20\"></path><path d=\"M 731.9 344.3 L 731.9 383.0 L 741.1 383.0\" id=\"21,19\"></path><path d=\"M 740.7 305.7 L 740.7 324.2 L 815.0 324.2\" id=\"20,18\"></path><path d=\"M 741.1 383.0 L 741.1 404.6 L 787.1 404.6\" id=\"19,17\"></path><path d=\"M 787.1 404.6 L 787.1 423.2 L 787.1 423.2\" id=\"17,16\"></path><path d=\"M 51.9 116.5 L 51.9 64.5 L 60.1 64.5\" id=\"30,15\"></path><path d=\"M 513.5 110.9 L 513.5 89.2 L 708.9 89.2\" id=\"28,14\"></path><path d=\"M 526.8 132.5 L 526.8 113.9 L 640.2 113.9\" id=\"26,13\"></path><path d=\"M 593.6 151.1 L 593.6 138.7 L 612.1 138.7\" id=\"24,12\"></path><path d=\"M 593.6 151.1 L 593.6 163.4 L 612.1 163.4\" id=\"24,11\"></path><path d=\"M 88.7 226.2 L 88.7 188.2 L 895.8 188.2\" id=\"27,10\"></path><path d=\"M 847.0 225.3 L 847.0 212.9 L 847.0 212.9\" id=\"23,9\"></path><path d=\"M 847.0 225.3 L 847.0 237.6 L 865.4 237.6\" id=\"23,8\"></path><path d=\"M 731.8 303.3 L 731.8 262.4 L 787.0 262.4\" id=\"22,7\"></path><path d=\"M 740.7 305.7 L 740.7 287.1 L 796.4 287.1\" id=\"20,6\"></path><path d=\"M 815.0 324.2 L 815.0 311.8 L 824.1 311.8\" id=\"18,5\"></path><path d=\"M 815.0 324.2 L 815.0 336.6 L 815.0 336.6\" id=\"18,4\"></path><path d=\"M 741.1 383.0 L 741.1 361.3 L 787.1 361.3\" id=\"19,3\"></path><path d=\"M 787.1 404.6 L 787.1 386.1 L 805.3 386.1\" id=\"17,2\"></path><path d=\"M 787.1 423.2 L 787.1 410.8 L 787.1 410.8\" id=\"16,1\"></path><path d=\"M 787.1 423.2 L 787.1 435.5 L 787.1 435.5\" id=\"16,0\"></path></g><g class=\"toytree-TipLabels\" style=\"fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:16px;font-weight:normal;stroke:none;white-space:pre\"><g transform=\"translate(787.07,435.53)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">19-7190</text></g><g transform=\"translate(787.07,410.79)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4948</text></g><g transform=\"translate(805.34,386.05)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">19-8206</text></g><g transform=\"translate(787.13,361.32)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">362p</text></g><g transform=\"translate(814.96,336.58)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1579</text></g><g transform=\"translate(824.09,311.84)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">2423p</text></g><g transform=\"translate(796.37,287.11)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1414</text></g><g transform=\"translate(787.05,262.37)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4973</text></g><g transform=\"translate(865.35,237.63)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">3825</text></g><g transform=\"translate(847.01,212.89)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">2014</text></g><g transform=\"translate(895.78,188.16)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1212</text></g><g transform=\"translate(612.10,163.42)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4798</text></g><g transform=\"translate(612.07,138.68)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">548</text></g><g transform=\"translate(640.25,113.95)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">3496</text></g><g transform=\"translate(708.89,89.21)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4960</text></g><g transform=\"translate(60.09,64.47)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">ref</text></g></g></g></g><g class=\"toyplot-coordinates-Axis\" id=\"td4c3bfb8cc8a41469380dbbc8f40ebc0\" transform=\"translate(50.0,450.0)translate(0,20.0)\"><line x1=\"1.8753438130323143\" y1=\"0\" x2=\"845.7800596776077\" y2=\"0\" style=\"\"></line><g><line x1=\"845.7800596776077\" y1=\"0\" x2=\"845.7800596776077\" y2=\"-5\" style=\"\"></line><line x1=\"752.0128690259882\" y1=\"0\" x2=\"752.0128690259882\" y2=\"-5\" style=\"\"></line><line x1=\"658.2456783743687\" y1=\"0\" x2=\"658.2456783743687\" y2=\"-5\" style=\"\"></line><line x1=\"564.4784877227492\" y1=\"0\" x2=\"564.4784877227492\" y2=\"-5\" style=\"\"></line><line x1=\"470.7112970711297\" y1=\"0\" x2=\"470.7112970711297\" y2=\"-5\" style=\"\"></line><line x1=\"376.9441064195102\" y1=\"0\" x2=\"376.9441064195102\" y2=\"-5\" style=\"\"></line><line x1=\"283.17691576789076\" y1=\"0\" x2=\"283.17691576789076\" y2=\"-5\" style=\"\"></line><line x1=\"189.4097251162713\" y1=\"0\" x2=\"189.4097251162713\" y2=\"-5\" style=\"\"></line><line x1=\"95.64253446465179\" y1=\"0\" x2=\"95.64253446465179\" y2=\"-5\" style=\"\"></line><line x1=\"1.8753438130323143\" y1=\"0\" x2=\"1.8753438130323143\" y2=\"-5\" style=\"\"></line></g><g><g transform=\"translate(845.7800596776077,6)\"><text x=\"-2.78\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">0</text></g><g transform=\"translate(752.0128690259882,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">10</text></g><g transform=\"translate(658.2456783743687,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">21</text></g><g transform=\"translate(564.4784877227492,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">31</text></g><g transform=\"translate(470.7112970711297,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">41</text></g><g transform=\"translate(376.9441064195102,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">51</text></g><g transform=\"translate(283.17691576789076,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">62</text></g><g transform=\"translate(189.4097251162713,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">72</text></g><g transform=\"translate(95.64253446465179,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">82</text></g><g transform=\"translate(1.8753438130323143,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">92</text></g></g><g class=\"toyplot-coordinates-Axis-coordinates\" style=\"visibility:hidden\" transform=\"\"><line x1=\"0\" x2=\"0\" y1=\"-3.0\" y2=\"4.5\" style=\"stroke:rgb(43.9%,50.2%,56.5%);stroke-opacity:1.0;stroke-width:1.0\"></line><text x=\"0\" y=\"-6\" style=\"alignment-baseline:alphabetic;fill:rgb(43.9%,50.2%,56.5%);fill-opacity:1.0;font-size:10px;font-weight:normal;stroke:none;text-anchor:middle\"></text></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "modules[\"toyplot/canvas/id\"] = \"t101fcba8df274600abf79d1f60a3f9de\";\n",
       "modules[\"toyplot/canvas\"] = (function(canvas_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + canvas_id);\n",
       "    })(modules[\"toyplot/canvas/id\"]);\n",
       "modules[\"toyplot.coordinates.Axis\"] = (\n",
       "        function(canvas)\n",
       "        {\n",
       "            function sign(x)\n",
       "            {\n",
       "                return x < 0 ? -1 : x > 0 ? 1 : 0;\n",
       "            }\n",
       "\n",
       "            function mix(a, b, amount)\n",
       "            {\n",
       "                return ((1.0 - amount) * a) + (amount * b);\n",
       "            }\n",
       "\n",
       "            function log(x, base)\n",
       "            {\n",
       "                return Math.log(Math.abs(x)) / Math.log(base);\n",
       "            }\n",
       "\n",
       "            function in_range(a, x, b)\n",
       "            {\n",
       "                var left = Math.min(a, b);\n",
       "                var right = Math.max(a, b);\n",
       "                return left <= x && x <= right;\n",
       "            }\n",
       "\n",
       "            function inside(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.min, range, segment.range.max))\n",
       "                        return true;\n",
       "                }\n",
       "                return false;\n",
       "            }\n",
       "\n",
       "            function to_domain(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.bounds.min, range, segment.range.bounds.max))\n",
       "                    {\n",
       "                        if(segment.scale == \"linear\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            return mix(segment.domain.min, segment.domain.max, amount)\n",
       "                        }\n",
       "                        else if(segment.scale[0] == \"log\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            var base = segment.scale[1];\n",
       "                            return sign(segment.domain.min) * Math.pow(base, mix(log(segment.domain.min, base), log(segment.domain.max, base), amount));\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            var axes = {};\n",
       "\n",
       "            function display_coordinates(e)\n",
       "            {\n",
       "                var current = canvas.createSVGPoint();\n",
       "                current.x = e.clientX;\n",
       "                current.y = e.clientY;\n",
       "\n",
       "                for(var axis_id in axes)\n",
       "                {\n",
       "                    var axis = document.querySelector(\"#\" + axis_id);\n",
       "                    var coordinates = axis.querySelector(\".toyplot-coordinates-Axis-coordinates\");\n",
       "                    if(coordinates)\n",
       "                    {\n",
       "                        var projection = axes[axis_id];\n",
       "                        var local = current.matrixTransform(axis.getScreenCTM().inverse());\n",
       "                        if(inside(local.x, projection))\n",
       "                        {\n",
       "                            var domain = to_domain(local.x, projection);\n",
       "                            coordinates.style.visibility = \"visible\";\n",
       "                            coordinates.setAttribute(\"transform\", \"translate(\" + local.x + \")\");\n",
       "                            var text = coordinates.querySelector(\"text\");\n",
       "                            text.textContent = domain.toFixed(2);\n",
       "                        }\n",
       "                        else\n",
       "                        {\n",
       "                            coordinates.style.visibility= \"hidden\";\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            canvas.addEventListener(\"click\", display_coordinates);\n",
       "\n",
       "            var module = {};\n",
       "            module.show_coordinates = function(axis_id, projection)\n",
       "            {\n",
       "                axes[axis_id] = projection;\n",
       "            }\n",
       "\n",
       "            return module;\n",
       "        })(modules[\"toyplot/canvas\"]);\n",
       "(function(axis, axis_id, projection)\n",
       "        {\n",
       "            axis.show_coordinates(axis_id, projection);\n",
       "        })(modules[\"toyplot.coordinates.Axis\"],\"td4c3bfb8cc8a41469380dbbc8f40ebc0\",[{\"domain\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 5.932936431644437, \"min\": -92.54822671111111}, \"range\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 900.0, \"min\": 0.0}, \"scale\": \"linear\"}]);\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import toytree\n",
    "style={\"font-size\": \"16px\"}\n",
    "treefile=cleaned+'/snipgenie_results/tree.newick'\n",
    "tre = toytree.tree(treefile)\n",
    "tre = tre.root('ref')\n",
    "tre=tre.drop_tips('2061')\n",
    "canvas,t,r=tre.draw(layout='r',tip_labels_style=style,width=1000,height=500,scalebar=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "b72cb609-ed94-4456-b35e-66b87605fddb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"toyplot\" id=\"t8434ecc495244184ba3cd799f2a5d2f0\" style=\"text-align:center\"><svg class=\"toyplot-canvas-Canvas\" xmlns:toyplot=\"http://www.sandia.gov/toyplot\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns=\"http://www.w3.org/2000/svg\" width=\"1000.0px\" height=\"500.0px\" viewBox=\"0 0 1000.0 500.0\" preserveAspectRatio=\"xMidYMid meet\" style=\"background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0\" id=\"t78f29c9b68b54caa859c33ef1f0a982b\"><g class=\"toyplot-coordinates-Cartesian\" id=\"t0aba9a9150634959bdd0217b4313a6e7\"><clipPath id=\"t0e4d0831805f415f897c9ca516cb9d53\"><rect x=\"30.0\" y=\"30.0\" width=\"940.0\" height=\"440.0\"></rect></clipPath><g clip-path=\"url(#t0e4d0831805f415f897c9ca516cb9d53)\"><g class=\"toytree-mark-Toytree\" id=\"t8dbb72ad88d24f8683f12f7af0248265\"><g class=\"toytree-Edges\" style=\"fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2\"><path d=\"M 51.9 116.5 L 51.9 168.5 L 87.5 168.5\" id=\"30,29\"></path><path d=\"M 87.5 168.5 L 87.5 110.9 L 440.4 110.9\" id=\"29,28\"></path><path d=\"M 87.5 168.5 L 87.5 226.2 L 111.4 226.2\" id=\"29,27\"></path><path d=\"M 440.4 110.9 L 440.4 132.5 L 463.3 132.5\" id=\"28,26\"></path><path d=\"M 111.4 226.2 L 111.4 264.3 L 127.0 264.3\" id=\"27,25\"></path><path d=\"M 463.3 132.5 L 463.3 151.1 L 542.1 151.1\" id=\"26,24\"></path><path d=\"M 127.0 264.3 L 127.0 225.3 L 745.7 225.3\" id=\"25,23\"></path><path d=\"M 127.0 264.3 L 127.0 303.3 L 644.1 303.3\" id=\"25,22\"></path><path d=\"M 644.1 303.3 L 644.1 344.3 L 644.2 344.3\" id=\"22,21\"></path><path d=\"M 644.2 344.3 L 644.2 305.7 L 659.9 305.7\" id=\"21,20\"></path><path d=\"M 644.2 344.3 L 644.2 383.0 L 652.0 383.0\" id=\"21,19\"></path><path d=\"M 659.9 305.7 L 659.9 324.2 L 730.8 324.2\" id=\"20,18\"></path><path d=\"M 652.0 383.0 L 652.0 404.6 L 707.2 404.6\" id=\"19,17\"></path><path d=\"M 707.2 404.6 L 707.2 423.2 L 707.2 423.2\" id=\"17,16\"></path><path d=\"M 51.9 116.5 L 51.9 64.5 L 87.5 64.5\" id=\"30,15\"></path><path d=\"M 440.4 110.9 L 440.4 89.2 L 607.5 89.2\" id=\"28,14\"></path><path d=\"M 463.3 132.5 L 463.3 113.9 L 566.0 113.9\" id=\"26,13\"></path><path d=\"M 542.1 151.1 L 542.1 138.7 L 557.9 138.7\" id=\"24,12\"></path><path d=\"M 542.1 151.1 L 542.1 163.4 L 557.9 163.4\" id=\"24,11\"></path><path d=\"M 111.4 226.2 L 111.4 188.2 L 607.6 188.2\" id=\"27,10\"></path><path d=\"M 745.7 225.3 L 745.7 212.9 L 753.6 212.9\" id=\"23,9\"></path><path d=\"M 745.7 225.3 L 745.7 237.6 L 895.8 237.6\" id=\"23,8\"></path><path d=\"M 644.1 303.3 L 644.1 262.4 L 778.3 262.4\" id=\"22,7\"></path><path d=\"M 659.9 305.7 L 659.9 287.1 L 715.1 287.1\" id=\"20,6\"></path><path d=\"M 730.8 324.2 L 730.8 311.8 L 738.7 311.8\" id=\"18,5\"></path><path d=\"M 730.8 324.2 L 730.8 336.6 L 888.9 336.6\" id=\"18,4\"></path><path d=\"M 652.0 383.0 L 652.0 361.3 L 699.3 361.3\" id=\"19,3\"></path><path d=\"M 707.2 404.6 L 707.2 386.1 L 809.8 386.1\" id=\"17,2\"></path><path d=\"M 707.2 423.2 L 707.2 410.8 L 707.2 410.8\" id=\"16,1\"></path><path d=\"M 707.2 423.2 L 707.2 435.5 L 730.8 435.5\" id=\"16,0\"></path></g><g class=\"toytree-TipLabels\" style=\"fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:16px;font-weight:normal;stroke:none;white-space:pre\"><g transform=\"translate(730.81,435.53)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">19-8206</text></g><g transform=\"translate(707.24,410.79)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">19-7190</text></g><g transform=\"translate(809.77,386.05)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4948</text></g><g transform=\"translate(699.28,361.32)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">362p</text></g><g transform=\"translate(888.86,336.58)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1579</text></g><g transform=\"translate(738.70,311.84)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">2423p</text></g><g transform=\"translate(715.12,287.11)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1414</text></g><g transform=\"translate(778.31,262.37)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4973</text></g><g transform=\"translate(895.78,237.63)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">2014</text></g><g transform=\"translate(753.59,212.89)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">3825</text></g><g transform=\"translate(607.64,188.16)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">1212</text></g><g transform=\"translate(557.94,163.42)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4798</text></g><g transform=\"translate(557.89,138.68)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">548</text></g><g transform=\"translate(565.99,113.95)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">3496</text></g><g transform=\"translate(607.47,89.21)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">4960</text></g><g transform=\"translate(87.48,64.47)rotate(0)\"><text x=\"15.00\" y=\"4.09\" style=\"\">ref</text></g></g></g></g><g class=\"toyplot-coordinates-Axis\" id=\"tba3b318d3f784e5598d85ec7045249d5\" transform=\"translate(50.0,450.0)translate(0,20.0)\"><line x1=\"1.8753438130322606\" y1=\"0\" x2=\"845.7800596776077\" y2=\"0\" style=\"\"></line><g><line x1=\"845.7800596776077\" y1=\"0\" x2=\"845.7800596776077\" y2=\"-5\" style=\"\"></line><line x1=\"752.012869025988\" y1=\"0\" x2=\"752.012869025988\" y2=\"-5\" style=\"\"></line><line x1=\"658.2456783743687\" y1=\"0\" x2=\"658.2456783743687\" y2=\"-5\" style=\"\"></line><line x1=\"564.4784877227492\" y1=\"0\" x2=\"564.4784877227492\" y2=\"-5\" style=\"\"></line><line x1=\"470.71129707112976\" y1=\"0\" x2=\"470.71129707112976\" y2=\"-5\" style=\"\"></line><line x1=\"376.9441064195103\" y1=\"0\" x2=\"376.9441064195103\" y2=\"-5\" style=\"\"></line><line x1=\"283.1769157678908\" y1=\"0\" x2=\"283.1769157678908\" y2=\"-5\" style=\"\"></line><line x1=\"189.4097251162714\" y1=\"0\" x2=\"189.4097251162714\" y2=\"-5\" style=\"\"></line><line x1=\"95.64253446465189\" y1=\"0\" x2=\"95.64253446465189\" y2=\"-5\" style=\"\"></line><line x1=\"1.875343813032372\" y1=\"0\" x2=\"1.875343813032372\" y2=\"-5\" style=\"\"></line></g><g><g transform=\"translate(845.7800596776077,6)\"><text x=\"-2.78\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">0</text></g><g transform=\"translate(752.012869025988,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">12</text></g><g transform=\"translate(658.2456783743687,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">24</text></g><g transform=\"translate(564.4784877227492,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">36</text></g><g transform=\"translate(470.71129707112976,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">48</text></g><g transform=\"translate(376.9441064195103,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">60</text></g><g transform=\"translate(283.1769157678908,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">72</text></g><g transform=\"translate(189.4097251162714,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">83</text></g><g transform=\"translate(95.64253446465189,6)\"><text x=\"-5.56\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">95</text></g><g transform=\"translate(1.875343813032372,6)\"><text x=\"-8.34\" y=\"8.555\" style=\"fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:helvetica;font-size:10.0px;font-weight:normal;stroke:none;vertical-align:baseline;white-space:pre\">107</text></g></g><g class=\"toyplot-coordinates-Axis-coordinates\" style=\"visibility:hidden\" transform=\"\"><line x1=\"0\" x2=\"0\" y1=\"-3.0\" y2=\"4.5\" style=\"stroke:rgb(43.9%,50.2%,56.5%);stroke-opacity:1.0;stroke-width:1.0\"></line><text x=\"0\" y=\"-6\" style=\"alignment-baseline:alphabetic;fill:rgb(43.9%,50.2%,56.5%);fill-opacity:1.0;font-size:10px;font-weight:normal;stroke:none;text-anchor:middle\"></text></g></g></g></svg><div class=\"toyplot-behavior\"><script>(function()\n",
       "{\n",
       "var modules={};\n",
       "modules[\"toyplot/canvas/id\"] = \"t78f29c9b68b54caa859c33ef1f0a982b\";\n",
       "modules[\"toyplot/canvas\"] = (function(canvas_id)\n",
       "    {\n",
       "        return document.querySelector(\"#\" + canvas_id);\n",
       "    })(modules[\"toyplot/canvas/id\"]);\n",
       "modules[\"toyplot.coordinates.Axis\"] = (\n",
       "        function(canvas)\n",
       "        {\n",
       "            function sign(x)\n",
       "            {\n",
       "                return x < 0 ? -1 : x > 0 ? 1 : 0;\n",
       "            }\n",
       "\n",
       "            function mix(a, b, amount)\n",
       "            {\n",
       "                return ((1.0 - amount) * a) + (amount * b);\n",
       "            }\n",
       "\n",
       "            function log(x, base)\n",
       "            {\n",
       "                return Math.log(Math.abs(x)) / Math.log(base);\n",
       "            }\n",
       "\n",
       "            function in_range(a, x, b)\n",
       "            {\n",
       "                var left = Math.min(a, b);\n",
       "                var right = Math.max(a, b);\n",
       "                return left <= x && x <= right;\n",
       "            }\n",
       "\n",
       "            function inside(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.min, range, segment.range.max))\n",
       "                        return true;\n",
       "                }\n",
       "                return false;\n",
       "            }\n",
       "\n",
       "            function to_domain(range, projection)\n",
       "            {\n",
       "                for(var i = 0; i != projection.length; ++i)\n",
       "                {\n",
       "                    var segment = projection[i];\n",
       "                    if(in_range(segment.range.bounds.min, range, segment.range.bounds.max))\n",
       "                    {\n",
       "                        if(segment.scale == \"linear\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            return mix(segment.domain.min, segment.domain.max, amount)\n",
       "                        }\n",
       "                        else if(segment.scale[0] == \"log\")\n",
       "                        {\n",
       "                            var amount = (range - segment.range.min) / (segment.range.max - segment.range.min);\n",
       "                            var base = segment.scale[1];\n",
       "                            return sign(segment.domain.min) * Math.pow(base, mix(log(segment.domain.min, base), log(segment.domain.max, base), amount));\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            var axes = {};\n",
       "\n",
       "            function display_coordinates(e)\n",
       "            {\n",
       "                var current = canvas.createSVGPoint();\n",
       "                current.x = e.clientX;\n",
       "                current.y = e.clientY;\n",
       "\n",
       "                for(var axis_id in axes)\n",
       "                {\n",
       "                    var axis = document.querySelector(\"#\" + axis_id);\n",
       "                    var coordinates = axis.querySelector(\".toyplot-coordinates-Axis-coordinates\");\n",
       "                    if(coordinates)\n",
       "                    {\n",
       "                        var projection = axes[axis_id];\n",
       "                        var local = current.matrixTransform(axis.getScreenCTM().inverse());\n",
       "                        if(inside(local.x, projection))\n",
       "                        {\n",
       "                            var domain = to_domain(local.x, projection);\n",
       "                            coordinates.style.visibility = \"visible\";\n",
       "                            coordinates.setAttribute(\"transform\", \"translate(\" + local.x + \")\");\n",
       "                            var text = coordinates.querySelector(\"text\");\n",
       "                            text.textContent = domain.toFixed(2);\n",
       "                        }\n",
       "                        else\n",
       "                        {\n",
       "                            coordinates.style.visibility= \"hidden\";\n",
       "                        }\n",
       "                    }\n",
       "                }\n",
       "            }\n",
       "\n",
       "            canvas.addEventListener(\"click\", display_coordinates);\n",
       "\n",
       "            var module = {};\n",
       "            module.show_coordinates = function(axis_id, projection)\n",
       "            {\n",
       "                axes[axis_id] = projection;\n",
       "            }\n",
       "\n",
       "            return module;\n",
       "        })(modules[\"toyplot/canvas\"]);\n",
       "(function(axis, axis_id, projection)\n",
       "        {\n",
       "            axis.show_coordinates(axis_id, projection);\n",
       "        })(modules[\"toyplot.coordinates.Axis\"],\"tba3b318d3f784e5598d85ec7045249d5\",[{\"domain\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 6.896806615111103, \"min\": -107.58369477777778}, \"range\": {\"bounds\": {\"max\": Infinity, \"min\": -Infinity}, \"max\": 900.0, \"min\": 0.0}, \"scale\": \"linear\"}]);\n",
       "})();</script></div></div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clnames=list(X.index)+['ref']\n",
    "treefile='/storage/btbgenie/monaghan/monaghan_results/tree.newick'\n",
    "tre = toytree.tree(treefile)\n",
    "drop=[n for n in tre.get_tip_labels() if n not in clnames]\n",
    "tre=tre.drop_tips(drop)\n",
    "tre=tre.drop_tips('2061')\n",
    "tre = tre.root('ref')\n",
    "canvas,t,r=tre.draw(layout='r',tip_labels_style=style,width=1000,height=500,scalebar=True)\n",
    "tre.write(cleaned+'/snipgenie_results/old.newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "f2330d53-1ad3-4806-a1cc-7f08ad747e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "muts1 = app.read_csq_file(cleaned+'/snipgenie_results/csq.tsv')\n",
    "muts2 = app.read_csq_file('/storage/btbgenie/monaghan/monaghan_results/csq.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d9331ff-3396-4b3a-88f8-4b5697621478",
   "metadata": {},
   "outputs": [],
   "source": [
    "x=pd.read_csv('/storage/btbgenie/monaghan/monaghan_results/csq.matrix',index_col=[0,1,2,3])\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "id": "345f8cd9-69fd-4735-a6db-211340974ea0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>sample</th>\n",
       "      <th>2</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>snp_type</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>aa</th>\n",
       "      <th>nuc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [1, sample, 2, chrom, start, snp_type, gene, locus_tag, strand, feature_type, aa, nuc]\n",
       "Index: []"
      ]
     },
     "execution_count": 360,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts2[muts2.gene=='RRL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aacbc85-5bb1-4f78-bb77-e8be91c9d7a3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
