{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d65f5a1e-65a6-44af-bc8e-d3b3b0eeb616",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,shutil,subprocess\n",
    "import glob\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 200)\n",
    "import pylab as plt\n",
    "import matplotlib as mpl\n",
    "sys.path.append('c:/Users/damien/projects/snipgenie/')\n",
    "from snipgenie import tools, aligners, app, trees, plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a19c5676-784b-4dad-95f9-3581642e0af9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following options were supplied\n",
      "-------\n",
      "threads : 4\n",
      "outdir : E:/test_results\n",
      "labelsep : _\n",
      "input : ['E:/test_files']\n",
      "species : Mbovis-AF212297\n",
      "get_stats : False\n",
      "overwrite : False\n",
      "custom_filters : True\n",
      "buildtree : True\n",
      "labelindex : 0\n",
      "trim : False\n",
      "unmapped : False\n",
      "quality : 25\n",
      "aligner : bwa\n",
      "platform : illumina\n",
      "filters : QUAL>=40 && FORMAT/DP>=30 && DP4>=4\n",
      "mask : None\n",
      "reference : None\n",
      "gb_file : None\n",
      "omit_samples : []\n",
      "bootstraps : 100\n",
      "\n",
      "4 samples were loaded:\n",
      "----------------------\n",
      "     sample                                          filename1                                          filename2                       name1                       name2  read_length\n",
      "0  13-11594  E:\\test_files\\13-11594_S85_L001-4_R1_001.fastq.gz  E:\\test_files\\13-11594_S85_L001-4_R2_001.fastq.gz  13-11594_S85_L001-4_R1_001  13-11594_S85_L001-4_R2_001          149\n",
      "1  15-11643  E:\\test_files\\15-11643_S67_L001-4_R1_001.fastq.gz  E:\\test_files\\15-11643_S67_L001-4_R2_001.fastq.gz  15-11643_S67_L001-4_R1_001  15-11643_S67_L001-4_R2_001          149\n",
      "2  17-11662  E:\\test_files\\17-11662_S11_L001-4_R1_001.fastq.gz  E:\\test_files\\17-11662_S11_L001-4_R2_001.fastq.gz  17-11662_S11_L001-4_R1_001  17-11662_S11_L001-4_R2_001          149\n",
      "3  19-11957  E:\\test_files\\19-11957_S38_L001-4_R1_001.fastq.gz  E:\\test_files\\19-11957_S38_L001-4_R2_001.fastq.gz  19-11957_S38_L001-4_R1_001  19-11957_S38_L001-4_R2_001          150\n",
      "\n",
      "building index\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bwa.exe index C:\\Users\\damien\\.config\\snipgenie\\genome\\Mbovis_AF212297.fa\n",
      "aligning files\n",
      "--------------\n",
      "Using reference genome: C:\\Users\\damien\\.config\\snipgenie\\genome\\Mbovis_AF212297.fa\n",
      "4/4 samples already aligned\n",
      "\n",
      "calling variants\n",
      "----------------\n",
      "E:/test_results\\raw.bcf already exists\n",
      "calling variants..\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe call --ploidy 1 -m -v -o E:/test_results\\calls.vcf E:/test_results\\raw.bcf\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe reheader --samples E:/test_results\\samples.txt -o C:\\Users\\damien\\AppData\\Local\\Temp\\calls.vcf E:/test_results\\calls.vcf\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe filter -i \"QUAL>=40 && FORMAT/DP>=30 && DP4>=4\" -o E:/test_results\\filtered.vcf.gz -O z E:/test_results\\calls.vcf\n",
      "splitting snps and indels..\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe view -v snps -o E:/test_results\\snps.vcf.gz -O z E:/test_results\\filtered.vcf.gz\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe view -v indels -o E:/test_results\\indels.vcf.gz -O z E:/test_results\\filtered.vcf.gz\n",
      "using mask bed file c:\\Users\\damien\\projects\\snipgenie\\snipgenie\\data\\Mbovis_AF212297_mask.bed\n",
      "189 sites\n",
      "found 7 sites in masked regions\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe view E:/test_results\\temp.vcf -O z -o E:/test_results\\snps.vcf.gz\n",
      "proximity filter removed 0/182 sites\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe view E:/test_results\\temp.vcf -O z -o E:/test_results\\snps.vcf.gz\n",
      "consequence calling..\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe csq -f C:\\Users\\damien\\.config\\snipgenie\\genome\\Mbovis_AF212297.fa -g E:/test_results\\Mbovis_AF212297.gb.gff E:/test_results\\snps.vcf.gz -Ot -o E:/test_results\\csq.tsv\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\bcftools.exe csq -f C:\\Users\\damien\\.config\\snipgenie\\genome\\Mbovis_AF212297.fa -g E:/test_results\\Mbovis_AF212297.gb.gff E:/test_results\\indels.vcf.gz -Ot -o E:/test_results\\csq_indels.tsv\n",
      "took 10.0 seconds\n",
      "E:/test_results\\snps.vcf.gz\n",
      "\n",
      "making SNP matrix\n",
      "-----------------\n",
      "used 21 sites for core snps\n",
      "0 sites with at least one missing sample\n",
      "161 uninformative sites\n",
      "\n",
      "Done. Sample summary:\n",
      "---------------------\n",
      "     sample                       name1                       name2  read_length                             bam_file\n",
      "0  13-11594  13-11594_S85_L001-4_R1_001  13-11594_S85_L001-4_R2_001          149  E:\\test_results\\mapped\\13-11594.bam\n",
      "1  15-11643  15-11643_S67_L001-4_R1_001  15-11643_S67_L001-4_R2_001          149  E:\\test_results\\mapped\\15-11643.bam\n",
      "2  17-11662  17-11662_S11_L001-4_R1_001  17-11662_S11_L001-4_R2_001          149  E:\\test_results\\mapped\\17-11662.bam\n",
      "3  19-11957  19-11957_S38_L001-4_R1_001  19-11957_S38_L001-4_R2_001          150  E:\\test_results\\mapped\\19-11957.bam\n",
      "\n",
      "building tree\n",
      "-------------\n",
      "C:\\Users\\damien\\.config\\snipgenie\\binaries\\RAxML.exe -f a -N 100 -T 4 -m GTRCAT -V -p 97820216 -x 51825884 -n variants -w E:\\test_results -s E:/test_results\\core.fa\n",
      "Error building tree. Is RAxML installed?\n"
     ]
    }
   ],
   "source": [
    "reload(app)\n",
    "reload(tools)\n",
    "reload(trees)\n",
    "args = {'threads':4, 'outdir': \"E:/test_results\", 'labelsep':'_',\n",
    "        'input':['E:/test_files'],        \n",
    "         'species': 'Mbovis-AF212297','get_stats':False,\n",
    "         'overwrite':False,\n",
    "         'custom_filters': True,\n",
    "         'buildtree':True}\n",
    "W = app.WorkFlow(**args)\n",
    "st = W.setup()\n",
    "W.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec82f550-ca99-4f9c-94fe-b19612125f2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33525d4e-9b5e-47cd-b828-f27388edb882",
   "metadata": {},
   "outputs": [],
   "source": [
    "'parallel bcftools mpileup -r {{1}} -a {a} -O b --min-MQ 60 -o {{2}} -f {r} {b} ::: {reg} :::+ {o}'\\\n",
    "                .format(r=ref, reg=regstr, b=bam_files, o=filesstr, a=annotatestr)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
