{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNP typing and lineage naming from WGS using reference population\n",
    "\n",
    "References:\n",
    "\n",
    "* https://www.nature.com/articles/ncomms5812\n",
    "* https://www.frontiersin.org/articles/10.3389/fmicb.2020.00843/full\n",
    "* https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3502966/\n",
    "\n",
    "Selection of reference population:\n",
    "\n",
    "* Global\n",
    "* European\n",
    "* UK/Ireland\n",
    "* All Ireland (republic + NI)\n",
    "\n",
    "required: \n",
    "\n",
    "* We need to be able to add new samples to the existing **reference** population phylogeny without having to re-analyse everything together.\n",
    "* We need to have a sufficiently sampled reference matrix that it will cover most potential new inputs. Otherwise they will be considered outgroups.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pylab as plt\n",
    "import scipy.cluster.hierarchy as shc\n",
    "from sklearn.preprocessing import normalize\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO,AlignIO\n",
    "from Bio import Phylo\n",
    "import seaborn as sns\n",
    "import toytree\n",
    "from snipgenie import app, trees, tools, snp_typing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [],
   "source": [
    "def snps_to_fasta(snpmat, outfile):\n",
    "    \"\"\"Write snp matrix to fasta file\"\"\"\n",
    "    \n",
    "    snpmat = snpmat.fillna('N')\n",
    "    recs = []\n",
    "    for col in snpmat.columns[1:]:\n",
    "        seq = ''.join(snpmat[col])        \n",
    "        seqrec = SeqRecord(Seq(seq),id=col)\n",
    "        recs.append(seqrec)\n",
    "    SeqIO.write(recs, outfile, 'fasta')\n",
    "    return    \n",
    "\n",
    "def tree_from_snps(snpmat):\n",
    "    snps_to_fasta(snpmat, 'snps.fa') \n",
    "    treefile = trees.run_fasttree('snps.fa')\n",
    "    tre = toytree.tree(treefile)\n",
    "    mystyle = { \"layout\": 'r','node_sizes':1,'tip_labels_align':False}\n",
    "    tre.ladderize().draw(**mystyle,width=700);\n",
    "    return tre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#snps_to_fasta(nucmat, 'snps.fa')\n",
    "tree_from_snps(nucmat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucmat = pd.read_csv('../snipgenie/data/nuc_snps_ireland.txt',sep=' ')\n",
    "nucmat = nucmat.set_index('pos')\n",
    "ref=nucmat['ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SequenceName</th>\n",
       "      <th>ClusterNumber</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>19-6359</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>19-7209</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>19-1603</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4794</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SequenceName  ClusterNumber\n",
       "10      19-6359              3\n",
       "11      19-7209              3\n",
       "12      19-1603              3\n",
       "13         4794              3"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusts=pd.read_csv('/storage/btbgenie/all_ireland_results/clusters.txt',sep='\\t')\n",
    "clusts[10:14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_ref_snps():\n",
    "    \n",
    "    nucmat = pd.read_csv('../snipgenie/data/nuc_snps_ireland.txt',sep=' ')\n",
    "    nucmat = nucmat.set_index('pos')\n",
    "    clusts=pd.read_csv('/storage/btbgenie/all_ireland_results/clusters.txt',sep='\\t')\n",
    "    X=nucmat.T.merge(clusts,left_index=True,right_on='SequenceName').set_index(['ClusterNumber']).T\n",
    "    return X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ClusterNumber</th>\n",
       "      <th>2</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>5</th>\n",
       "      <th>-1</th>\n",
       "      <th>5</th>\n",
       "      <th>...</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1303</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1456</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 249 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ClusterNumber  2  5  5  5  5  5  5  5 -1  5  ...  7  7  7  7  7  7  7  7  7  7\n",
       "687            C  C  C  C  C  C  C  C  C  C  ...  C  C  C  C  C  C  C  C  C  C\n",
       "937            A  A  A  A  A  A  A  A  A  A  ...  A  A  A  A  A  A  A  A  A  A\n",
       "1303           G  G  G  G  G  G  G  G  G  G  ...  G  G  G  G  G  G  G  G  G  G\n",
       "1456           G  G  G  G  G  G  G  G  G  G  ...  G  G  G  G  G  G  G  G  G  G\n",
       "\n",
       "[4 rows x 249 columns]"
      ]
     },
     "execution_count": 508,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = make_ref_snps()\n",
    "X[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print (X.loc[:,3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## find snps unique to clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     clade      pos allele\n",
      "0        2   490878      G\n",
      "1        2   804997      T\n",
      "2        2   941068      A\n",
      "3        2  1124266      G\n",
      "4        2  1442194      T\n",
      "..     ...      ...    ...\n",
      "229      8  2944945      T\n",
      "230      8  3191587      T\n",
      "231      8  3256582      A\n",
      "232      8  3440364      A\n",
      "233      8  3772840      C\n",
      "\n",
      "[234 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "def get_clade_snps(refmat):\n",
    "    \"\"\"get unique clade snps from a snp matrix\n",
    "       returns: a dataframe with unique positions/allele for each clade\n",
    "       with this format\n",
    "              clade      pos allele\n",
    "           2   490878      G\n",
    "           2   804997      T\n",
    "           2   941068      A\n",
    "           2  1124266      G\n",
    "    \"\"\"\n",
    "    \n",
    "    res=[]\n",
    "    clusters = refmat.columns.unique()\n",
    "    for c in clusters:\n",
    "        for pos,r in list(refmat.iterrows()):\n",
    "            #print (pos)\n",
    "            a = r[c]\n",
    "            b = r[~r.index.isin([c])]\n",
    "            #print (len(r),len(a),len(b))\n",
    "            f1 = a.value_counts()\n",
    "            f2 = b.value_counts()\n",
    "            alt1 = f1.index[0]\n",
    "            if len(f1)>1:\n",
    "                continue\n",
    "            alt2 = f2.index[0]\n",
    "            if alt1 in f2:\n",
    "                continue\n",
    "            #print (f1,alt2)\n",
    "            #print (f1,f2)\n",
    "            res.append((c,pos,alt1))\n",
    "    \n",
    "    res = pd.DataFrame(res,columns=['clade','pos','allele'])\n",
    "    print (res)\n",
    "    return res\n",
    "\n",
    "refmat = make_ref_snps()\n",
    "res = get_clade_snps(refmat)\n",
    "\n",
    "res.to_csv('clade_snps.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ClusterNumber\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "    ..\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "6    C\n",
       "Name: 51709, Length: 72, dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check results\n",
    "X.loc[51709,6]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## identify sample from clade-specific SNPs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 634,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{6}"
      ]
     },
     "execution_count": 634,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = '1579'\n",
    "\n",
    "def lookup_sample(snptable, snps):\n",
    "    \"\"\"Look up a sample using snps and known clades\n",
    "        snptable: reference lookup table\n",
    "        snps: a series with snps at each position for the\n",
    "        given sample, this can be derived from a single row\n",
    "        in the snp matrix produced from snipgenie\n",
    "    \"\"\"     \n",
    "        \n",
    "    found=[]\n",
    "    for i,r in snptable.iterrows():\n",
    "        if not r.pos in snps.index:\n",
    "            continue\n",
    "        if snps[r.pos] == r.allele:\n",
    "            #print (r.pos,r.allele,r.clade)\n",
    "            found.append(r.clade)\n",
    "    if len(found) == 0:\n",
    "        return\n",
    "    return set(found)\n",
    " \n",
    "reload(snp_typing)\n",
    "\n",
    "x = nucmat.set_index('pos')[sample]\n",
    "snptable = snp_typing.clade_snps\n",
    "lookup_sample(snptable, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 636,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref {2}\n",
      "31-12952 {5}\n",
      "48-MBovis {5}\n"
     ]
    }
   ],
   "source": [
    "def type_samples(nucmat):\n",
    "    \"\"\"\n",
    "    Type multiple samples.\n",
    "    Args:\n",
    "        nucmat: a dataframe with the following format-\n",
    "        pos       687  937  1303 ..      \n",
    "        sample1    C    A    G \n",
    "        sample2    C    A    G\n",
    "        ...\n",
    "    Returns:\n",
    "        types for each sample\n",
    "    \"\"\"\n",
    "    snptable = snp_typing.clade_snps\n",
    "    for name,r in nucmat.iterrows():\n",
    "        #print (r)\n",
    "        cl = lookup_sample(snptable, r)\n",
    "        print (name,cl)\n",
    "        \n",
    "snps = nucmat.set_index('pos').T[:3]\n",
    "type_samples(snps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test on isolated subset of samples - must use uninformative snps from subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 642,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 191 sites\n",
      "0 sites with at least one missing sample\n",
      "0 uninformative sites\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>pos</th>\n",
       "      <th>33788</th>\n",
       "      <th>41437</th>\n",
       "      <th>69913</th>\n",
       "      <th>130237</th>\n",
       "      <th>160535</th>\n",
       "      <th>166696</th>\n",
       "      <th>173274</th>\n",
       "      <th>229412</th>\n",
       "      <th>232188</th>\n",
       "      <th>246207</th>\n",
       "      <th>...</th>\n",
       "      <th>4130927</th>\n",
       "      <th>4133879</th>\n",
       "      <th>4162554</th>\n",
       "      <th>4180986</th>\n",
       "      <th>4191866</th>\n",
       "      <th>4216874</th>\n",
       "      <th>4217177</th>\n",
       "      <th>4227256</th>\n",
       "      <th>4298265</th>\n",
       "      <th>4311425</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17-11662</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15-11643</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19-11957</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13-11594</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 191 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "pos      33788   41437   69913   130237  160535  166696  173274  229412   \\\n",
       "ref            A       T       T       T       T       A       G       C   \n",
       "17-11662       G       C       C       C       C       G       T       C   \n",
       "15-11643       G       C       C       C       C       G       T       C   \n",
       "19-11957       G       C       C       C       C       G       T       C   \n",
       "13-11594       G       C       C       C       C       G       T       G   \n",
       "\n",
       "pos      232188  246207   ... 4130927 4133879 4162554 4180986 4191866 4216874  \\\n",
       "ref            G       C  ...       A       C       T       A       G       T   \n",
       "17-11662       C       T  ...       C       T       C       G       A       C   \n",
       "15-11643       C       T  ...       C       T       C       G       A       C   \n",
       "19-11957       C       T  ...       C       T       C       G       A       C   \n",
       "13-11594       C       T  ...       C       T       C       G       A       C   \n",
       "\n",
       "pos      4217177 4227256 4298265 4311425  \n",
       "ref            C       G       T       C  \n",
       "17-11662       C       G       C       T  \n",
       "15-11643       T       A       C       T  \n",
       "19-11957       T       A       C       T  \n",
       "13-11594       C       G       C       T  \n",
       "\n",
       "[5 rows x 191 columns]"
      ]
     },
     "execution_count": 642,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file='../test_results/filtered.vcf.gz'\n",
    "snprecs, testmat = tools.fasta_alignment_from_vcf(vcf_file, uninformative=True)\n",
    "testmat.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 643,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref {2}\n",
      "17-11662 {5}\n",
      "15-11643 {5}\n",
      "19-11957 {5}\n",
      "13-11594 {5}\n"
     ]
    }
   ],
   "source": [
    "type_samples(testmat.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## store snp data to database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## encode snps to store in a single field in DB?\n",
    "\n",
    "e.g. string encoding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos\n",
       "687        C\n",
       "937        A\n",
       "1303       G\n",
       "1456       G\n",
       "1584       T\n",
       "          ..\n",
       "3832251    G\n",
       "3833590    G\n",
       "3835182    G\n",
       "3839650    T\n",
       "3842525    G\n",
       "Length: 1746, dtype: object"
      ]
     },
     "execution_count": 602,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x=snps.iloc[0]\n",
    "\n",
    "def encode_snps(x):\n",
    "    \"\"\"encode snps as string for storage\"\"\"\n",
    "    \n",
    "    s=[]\n",
    "    for i in zip(x.index.astype(str),x.values):\n",
    "        s.append(''.join(i))\n",
    "    s = ';'.join(s)\n",
    "    return s\n",
    "\n",
    "s = encode_snps(x)\n",
    "\n",
    "import re\n",
    "def decode_snps(s):\n",
    "    \"\"\"decode snps\"\"\"\n",
    "    \n",
    "    x=s.split(';')\n",
    "    pos=[]\n",
    "    alleles=[]\n",
    "    for i in x:\n",
    "        n,p,a = re.split(r'(\\d+)', i)\n",
    "        pos.append(p)\n",
    "        alleles.append(a)\n",
    "    x = pd.Series(alleles,pos)\n",
    "    x.index.name='pos'\n",
    "    return x\n",
    "\n",
    "x = decode_snps(s)\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>pos</th>\n",
       "      <th>687</th>\n",
       "      <th>937</th>\n",
       "      <th>1303</th>\n",
       "      <th>1456</th>\n",
       "      <th>1584</th>\n",
       "      <th>2532</th>\n",
       "      <th>4480</th>\n",
       "      <th>8048</th>\n",
       "      <th>8150</th>\n",
       "      <th>8741</th>\n",
       "      <th>...</th>\n",
       "      <th>3823420</th>\n",
       "      <th>3825991</th>\n",
       "      <th>3827894</th>\n",
       "      <th>3828728</th>\n",
       "      <th>3830688</th>\n",
       "      <th>3832251</th>\n",
       "      <th>3833590</th>\n",
       "      <th>3835182</th>\n",
       "      <th>3839650</th>\n",
       "      <th>3842525</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31-12952</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48-MBovis</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49-MBovis</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28-12935</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1746 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "pos       687     937     1303    1456    1584    2532    4480    8048     \\\n",
       "ref             C       A       G       G       T       C       T       C   \n",
       "31-12952        C       A       G       G       T       C       T       C   \n",
       "48-MBovis       C       A       G       G       T       C       T       C   \n",
       "49-MBovis       C       A       G       G       T       C       T       C   \n",
       "28-12935        C       A       G       G       T       C       T       C   \n",
       "\n",
       "pos       8150    8741     ... 3823420 3825991 3827894 3828728 3830688  \\\n",
       "ref             C       T  ...       T       G       T       G       G   \n",
       "31-12952        C       C  ...       T       G       T       G       G   \n",
       "48-MBovis       C       C  ...       T       G       T       G       G   \n",
       "49-MBovis       C       C  ...       T       G       T       G       G   \n",
       "28-12935        C       C  ...       T       G       T       G       G   \n",
       "\n",
       "pos       3832251 3833590 3835182 3839650 3842525  \n",
       "ref             G       G       G       T       G  \n",
       "31-12952        G       G       G       C       G  \n",
       "48-MBovis       G       G       G       C       G  \n",
       "49-MBovis       G       G       G       C       G  \n",
       "28-12935        G       G       G       C       G  \n",
       "\n",
       "[5 rows x 1746 columns]"
      ]
     },
     "execution_count": 524,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d=nucmat.set_index('pos').T[:5]\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b10a8db164e0754105b7a99be72e3fe5\n"
     ]
    }
   ],
   "source": [
    "#assign ID\n",
    "import uuid\n",
    "  \n",
    "id = uuid.uuid1()\n",
    "id\n",
    "import hashlib\n",
    "hash_object = hashlib.md5(b'Hello World')\n",
    "print(hash_object.hexdigest())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isolate: 42343, SNP-type: None"
      ]
     },
     "execution_count": 611,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class isolate():\n",
    "    \n",
    "    def __init__(self, id, ):\n",
    "        \n",
    "        self.id = id\n",
    "        self.clade = None\n",
    "        self.species = 'M.bovis'\n",
    "        return\n",
    "    \n",
    "    def get_type(self):\n",
    "        \n",
    "        return\n",
    "    \n",
    "    def get_record(self):\n",
    "        \"\"\"Create isolate record from SNP typing\"\"\"\n",
    "\n",
    "        return\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return 'isolate: %s, SNP-type: %s' %(self.id, self.clade)\n",
    "\n",
    "iso = isolate('42343')\n",
    "iso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
