{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## snipgenie testing notebook\n",
    "\n",
    "Updated files can be found here: https://github.com/dmnfarrell/gordon-group/tree/master/mbovis_annotation\n",
    "\n",
    "links:\n",
    "\n",
    "* https://github.com/JosephCrispell/GeneralTools/tree/master/ProcessingPipeline\n",
    "* [Woodchester Park eLife paper](https://elifesciences.org/articles/45833)\n",
    "* [An African origin for Mycobacterium bovis](https://academic.oup.com/emph/article/2020/1/49/5719036)\n",
    "* http://www.htslib.org/doc/bcftools.html\n",
    "* https://mtbgenomicsworkshop.readthedocs.io/\n",
    "* [BCFtools cheat sheet](https://gist.github.com/elowy01/93922762e131d7abd3c7e8e166a74a0b)\n",
    "* https://biopython.org/wiki/Category:Cookbook\n",
    "* https://pyvcf.readthedocs.io/en/v0.4.6/INTRO.html\n",
    "* https://gatkforums.broadinstitute.org/gatk/discussion/1268/what-is-a-vcf-and-how-should-i-interpret-it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,shutil,subprocess\n",
    "import random\n",
    "import glob, time\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.width', 200)\n",
    "import pylab as plt\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from gzip import open as gzopen\n",
    "sys.path.append('pathogenie')\n",
    "from snipgenie import tools, aligners, app, trees, plotting\n",
    "#from matplotlib_venn import venn2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mbovis samples from SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sra = pd.read_csv('../sra_mbovis_runtable.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get master table with Loiseau et al. meta data¶"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('../Loiseau_TableS1.xlsx')\n",
    "master = pd.merge(df,sra,left_on='BIOSAMPLE',right_on='BioSample',how='inner')\n",
    "master.to_csv('../mbovis_sra_master.csv',index=False)\n",
    "cols = ['Run','CLONAL_COMPLEX','COUNTRY_ISOLATION','LibraryLayout','Host','ReleaseDate','Bytes','Bases','GENOME_COVERAGE']\n",
    "#master.columns\n",
    "#master.geo_loc_name_country.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Representative samples of lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Run CLONAL_COMPLEX COUNTRY_ISOLATION LibraryLayout        Host           ReleaseDate        Bytes        Bases  GENOME_COVERAGE\n",
      "36     ERR841808            Eu1                UK        PAIRED         Cow  2015-12-02T00:00:00Z  130651179.0  196040478.0            39.78\n",
      "245   SRR5216872            Eu1       New Zealand        PAIRED      BOVINE  2017-01-31T00:00:00Z  194523112.0  295803584.0            41.39\n",
      "464   SRR5486071       Unknown2          Ethiopia        PAIRED  Bos taurus  2017-04-28T00:00:00Z  251875435.0  445066100.0            50.71\n",
      "677    ERR125601            Eu1           Ireland        PAIRED         NaN  2012-10-01T00:00:00Z  277182504.0  346340680.0            68.83\n",
      "765   ERR2815558            Af1               NaN        PAIRED         NaN  2018-09-29T00:00:00Z  162711425.0  367702705.0            67.64\n",
      "886   SRR1791768            Eu2            Canada        PAIRED      Cattle  2015-03-11T00:00:00Z  340596851.0  450406275.0            37.69\n",
      "1076  SRR1791960            Eu2            Mexico        PAIRED      Cattle  2015-03-11T00:00:00Z  234700051.0  344439293.0            38.20\n",
      "1118  SRR1792002            Eu2               USA        PAIRED      Cattle  2015-03-11T00:00:00Z  155793754.0  221710448.0            37.69\n",
      "2577  SRR8065079            Af2            Uganda        PAIRED  Bos taurus  2018-12-01T00:00:00Z  149270806.0  248725188.0            32.01\n",
      "2597  SRR8063654            Af2          Ethiopia        PAIRED  Bos taurus  2018-12-01T00:00:00Z  128374182.0  220964011.0            25.87\n",
      "2608  SRR8063665            Af2          Ethiopia        PAIRED  Bos taurus  2018-12-01T00:00:00Z  127610605.0  219741248.0            27.41\n"
     ]
    }
   ],
   "source": [
    "sub=master[(master.LibraryLayout=='PAIRED') & (master.COUNTRY_ISOLATION.notnull()) & (master.CLONAL_COMPLEX=='Eu2')]\n",
    "#print (sub[cols])\n",
    "#sub[cols].to_csv('../temp.csv')\n",
    "accessions=['SRR1792002','SRR5216872','SRR5486071','ERR841808','ERR125601',\n",
    "            'SRR8063654','SRR8063665','SRR8065079','SRR1791768','SRR1791960','ERR2815558']\n",
    "testset = master[master.Run.isin(accessions)]\n",
    "print (testset[cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fetch from SRA\n",
    "def fetch_sra_reads(df,path):\n",
    "    \"\"\"download a subset of reads \"\"\"\n",
    "    \n",
    "    for i,r in df.iterrows():\n",
    "        files = glob.glob(os.path.join(path,r.Run+'*'))        \n",
    "        if len(files)==0:\n",
    "            cmd = 'fastq-dump --split-files {n} --outdir {o}'.format(n=r.Run,o=path)\n",
    "            print (cmd)\n",
    "            subprocess.check_output(cmd,shell=True)\n",
    "\n",
    "fetch_sra_reads(testset,path='/storage/btbgenie/mbovis_sra')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blasting to known contaminants\n",
    "\n",
    "* https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2684-x#Sec2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get unmapped reads from a bam file\n",
    "\n",
    "infile='mapped/3_S57_L001.bam'\n",
    "cmd = 'samtools view -b -f 4 {i} > unmapped.bam'.format(i=infile)\n",
    "print (cmd)\n",
    "#align with bwa to known index of contaminants\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examine FASTQ file quality and define trimming parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(tools)\n",
    "\n",
    "df = tools.fastq_to_dataframe(testfile)\n",
    "df.length.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_reads_default(filename, outfile, right_quality=30):\n",
    "    \"\"\"Trim adapters\"\"\"\n",
    "        \n",
    "    fastq_parser = SeqIO.parse(gzopen(filename, \"rt\"), \"fastq\")\n",
    "    c=0\n",
    "    out = gzopen(outfile, \"wt\")\n",
    "    for record in fastq_parser:\n",
    "        score = record.letter_annotations[\"phred_quality\"]         \n",
    "        for i in range(len(score)-1,0,-1):\n",
    "            if score[i] >= right_quality:\n",
    "                break\n",
    "        #trimmed.append(record[:i])\n",
    "        #print (record[:i])        \n",
    "        #c+=1\n",
    "        #if c>100:\n",
    "        #    break    \n",
    "        SeqIO.write(record[:i],out,'fastq')\n",
    "    return\n",
    "\n",
    "testfile = 'mbovis_sra/SRR1791711_1.fastq.gz'\n",
    "st=time.time()\n",
    "trim_reads_default(testfile,'trimmed.fastq.gz')\n",
    "print (time.time()-st)\n",
    "df = tools.fastq_to_dataframe('trimmed.fastq.gz')\n",
    "df.length.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## multiprocess trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_fastq(filename, chunks=4):\n",
    "    \"\"\"Split fastq file\"\"\"\n",
    "    \n",
    "    names=[]\n",
    "    tmp = '/tmp'\n",
    "    tmpfiles = []\n",
    "    record_iter = SeqIO.parse(gzopen(filename, \"rt\"), \"fastq\")\n",
    "    length = \n",
    "    for i, batch in enumerate(tools.batch_iterator(record_iter, chunksize)):\n",
    "        tempfile = \"group_%i.fastq\" % (i + 1)\n",
    "        tmpfiles.append(tempfile)\n",
    "        names.append(tempfile)\n",
    "        with open(tempfile, \"w\") as handle:\n",
    "            count = SeqIO.write(batch, handle, \"fastq\")\n",
    "        print(\"Wrote %i records to %s\" % (count, tempfile))\n",
    "    return names\n",
    "\n",
    "split_fastq(testfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools.plot_fastq_qualities(testfile,limit=100000)\n",
    "plt.savefig('fastq_quals.png',dpi=100)\n",
    "tools.plot_fastq_gc_content(testfile, limit=50000)\n",
    "plt.savefig('gc.png',dpi=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get sample ids from fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path1='/storage/btbgenie/albania'\n",
    "files = glob.glob(os.path.join(path1,'*.fastq.gz'))\n",
    "path2='/storage/btbgenie/mbovis_sra'\n",
    "files = glob.glob(os.path.join(path2,'*.fastq.gz'))\n",
    "#add outgroup file\n",
    "#files.append\n",
    "files = app.get_files_from_paths([path1,path2])\n",
    "\n",
    "df = app.get_samples(files)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align the FASTQ files against reference and create VCF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(aligners)\n",
    "ref = app.mbovis_genome\n",
    "samples = app.get_samples(files)\n",
    "app.align_reads(samples, idx=ref, threads=10, overwrite=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## remove duplicated reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'samtools rmdup A_reads.bt2.sorted.bam A_reads.bt2.sorted.noDups.bam'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Keeping unmapped reads\n",
    "\n",
    "https://gist.github.com/davfre/8596159"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## variant calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[      1  434991  869981 1304971 1739962 2174952 2609942 3044933 3479923\n",
      " 3914913 4349904]\n",
      "parallel bcftools mpileup -r {1} -a \"AD,ADF,ADR,DP,SP,INFO/AD,INFO/ADF,INFO/ADR\" -O b  --min-MQ 60 -o {2} -f /home/damien/.config/snpgenie/genome/Mbovis_AF212297.fa mapped/ERR551704.bam mapped/6_S7_L001.bam mapped/3_S57_L001.bam ::: \"LT708304.1\":1-434990 \"LT708304.1\":434991-869980 \"LT708304.1\":869981-1304970 \"LT708304.1\":1304971-1739961 \"LT708304.1\":1739962-2174951 \"LT708304.1\":2174952-2609941 \"LT708304.1\":2609942-3044932 \"LT708304.1\":3044933-3479922 \"LT708304.1\":3479923-3914912 \"LT708304.1\":3914913-4349903 :::+ /tmp/1.bcf /tmp/434991.bcf /tmp/869981.bcf /tmp/1304971.bcf /tmp/1739962.bcf /tmp/2174952.bcf /tmp/2609942.bcf /tmp/3044933.bcf /tmp/3479923.bcf /tmp/3914913.bcf\n",
      "bcftools concat /tmp/1.bcf /tmp/434991.bcf /tmp/869981.bcf /tmp/1304971.bcf /tmp/1739962.bcf /tmp/2174952.bcf /tmp/2609942.bcf /tmp/3044933.bcf /tmp/3479923.bcf /tmp/3914913.bcf -O b -o mapped/raw.bcf\n",
      "bcftools call -V indels --ploidy 1 -m -v -o mapped/calls.vcf mapped/raw.bcf\n",
      "bcftools reheader --samples mapped/samples.txt -o /tmp/calls.vcf mapped/calls.vcf\n",
      "bcftools filter -i \"QUAL>=40 && FORMAT/DP>=30 && DP4>=4 && MQ>30\" -o mapped/filtered.vcf.gz -O z mapped/calls.vcf\n",
      "took 33.0 seconds\n",
      "33.45631980895996\n"
     ]
    }
   ],
   "source": [
    "reload(app)\n",
    "ref = app.mbovis_genome\n",
    "bam_files = glob.glob('mapped/*.bam')\n",
    "st=time.time()\n",
    "vcf_file = app.variant_calling(bam_files, ref, 'mapped', threads=10, overwrite=True)\n",
    "print (time.time()-st)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## freebayes calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['mapped/ERR551704.bam', 'mapped/6_S7_L001.bam', 'mapped/3_S57_L001.bam']\n",
      "freebayes -p 2 -P 0 -C 2 -F 0.05 --min-coverage 10 --min-repeat-entropy 1.0 -q 40 -m 20 --strict-vcf -f /home/damien/.config/snpgenie/genome/Mbovis_AF212297.fa -L files.txt > mapped/freebayes.vcf\n"
     ]
    }
   ],
   "source": [
    "fbcmd = 'freebayes'\n",
    "\n",
    "ref=app.mbovis_genome\n",
    "bam_files = glob.glob('mapped/*.bam')\n",
    "print (bam_files)\n",
    "out='mapped/freebayes.vcf'\n",
    "cmd = '{fb} -p 2 -P 0 -C 2 -F 0.05 --min-coverage 10'\\\n",
    "         ' --min-repeat-entropy 1.0 -q 40 -m 20 --strict-vcf'\\\n",
    "        ' -f {r} -L files.txt > {o}'.format(r=ref,o=out,fb=fbcmd)        \n",
    "print (cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "v1 = vcf_to_dataframe('result.vcf')\n",
    "print (v1.var_type.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "v2 = vcf_to_dataframe('mapped/raw.bcf')\n",
    "print (v2.var_type.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## filtering\n",
    "\n",
    "```\n",
    "FORMAT/AD   .. Allelic depth (Number=R,Type=Integer)\n",
    "FORMAT/ADF  .. Allelic depths on the forward strand (Number=R,Type=Integer)\n",
    "FORMAT/ADR  .. Allelic depths on the reverse strand (Number=R,Type=Integer)\n",
    "FORMAT/DP   .. Number of high-quality bases (Number=1,Type=Integer)\n",
    "FORMAT/SP   .. Phred-scaled strand bias P-value (Number=1,Type=Integer)\n",
    "FORMAT/SCR  .. Number of soft-clipped reads (Number=1,Type=Integer)\n",
    "\n",
    "INFO/AD     .. Total allelic depth (Number=R,Type=Integer)\n",
    "INFO/ADF    .. Total allelic depths on the forward strand (Number=R,Type=Integer)\n",
    "INFO/ADR    .. Total allelic depths on the reverse strand (Number=R,Type=Integer)\n",
    "INFO/SCR    .. Number of soft-clipped reads (Number=1,Type=Integer)\n",
    "\n",
    "FORMAT/DV   .. Deprecated in favor of FORMAT/AD; Number of high-quality non-reference bases, (Number=1,Type=Integer)\n",
    "FORMAT/DP4  .. Deprecated in favor of FORMAT/ADF and FORMAT/ADR; Number of high-quality ref-forward, ref-reverse,\n",
    "               alt-forward and alt-reverse bases (Number=4,Type=Integer)\n",
    "FORMAT/DPR  .. Deprecated in favor of FORMAT/AD; Number of high-quality bases for each observed allele (Number=R,Type=Integer)\n",
    "INFO/DPR    .. Deprecated in favor of INFO/AD; Number of high-quality bases for each observed allele (Number=R,Type=Integer)\n",
    "```\n",
    "\n",
    "Joe filters: DP - 30, DP4 - 4, MQ - 35, SUP - 0.95 and COV - 0.05. GQ and FQ = 0\n",
    "\n",
    "## Other options\n",
    "\n",
    "```\n",
    "biallelic: -m2 -M2\n",
    "uninformative sites, where the minor allele is present in at least one individual: --min-ac 1:minor\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "vcf_reader = vcf.Reader(open('mapped/filtered.vcf.gz', 'rb')) \n",
    "for record in vcf_reader:\n",
    "    print (record)\n",
    "    for sample in record.samples:\n",
    "        print (sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = 'bcftools filter -e \"QUAL<=40 && DP>=30 && DP4>=4 && MQ>=35\" -o filtered.vcf.gz -O z mapped/calls.vcf'\n",
    "print (cmd)\n",
    "tmp = subprocess.check_output(cmd,shell=True)\n",
    "v = vcf_to_dataframe('filtered.vcf.gz')\n",
    "v.var_type.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## consequence calling for vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools csq -f /home/damien/.config/snpgenie/genome/Mbovis_AF212297.fa -g ../snpgenie/data/Mbovis_csq_format.gff mapped/filtered.vcf.gz -Ot -o mapped/csq.tsv\n"
     ]
    }
   ],
   "source": [
    "reload(app)\n",
    "ref=app.mbovis_genome\n",
    "gff = '../snpgenie/data/Mbovis_csq_format.gff'\n",
    "out = 'mapped/csq.tsv'\n",
    "def csq_call():\n",
    "    cmd = 'bcftools csq -f {r} -g {g} mapped/filtered.vcf.gz -Ot -o {o}'.format(r=ref,g=gff,o=out)\n",
    "    print (cmd)\n",
    "    tmp = subprocess.check_output(cmd,shell=True)\n",
    "    \n",
    "csq_call()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(app)\n",
    "def read_csq_file(filename):\n",
    "    cols = ['1','filename','2','chrom','start','snp_type','gene','locus_tag','strand','feature_type','aa','nuc',]\n",
    "    csqdf = pd.read_csv(filename,sep='[|\\t]',comment='#',names=cols, engine='python')\n",
    "    return csqdf\n",
    "csq = read_csq_file('csqout.tsv')\n",
    "csq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vcf_to_dataframe(vcf_file):\n",
    "    \"\"\"\n",
    "    Convert a multi sample vcf to dataframe. Records each samples FORMAT fields.\n",
    "    Args:\n",
    "        vcf_file: input multi sample vcf\n",
    "    Returns: pandas DataFrame\n",
    "    \"\"\"\n",
    "\n",
    "    import vcf\n",
    "    ext = os.path.splitext(vcf_file)[1]\n",
    "    if ext == '.gz':\n",
    "        file = gzopen(vcf_file, \"rt\")\n",
    "    else:\n",
    "        file = open(vcf_file)\n",
    "    vcf_reader = vcf.Reader(file,'r')\n",
    "    res=[]\n",
    "    cols = ['sample','REF','ALT','mut','DP','ADF','ADR','AD','chrom','var_type',\n",
    "            'sub_type','start','end','QUAL']\n",
    "    i=0\n",
    "    for rec in vcf_reader:\n",
    "        #if i>50:\n",
    "        #    break\n",
    "        x = [rec.CHROM, rec.var_type, rec.var_subtype, rec.start, rec.end, rec.QUAL]\n",
    "        for sample in rec.samples:\n",
    "            if sample.gt_bases == None:\n",
    "                mut=''\n",
    "                row = [sample.sample, rec.REF, sample.gt_bases, mut, 0,0,0,0]\n",
    "            elif rec.REF != sample.gt_bases:\n",
    "                mut = str(rec.end)+rec.REF+'>'+sample.gt_bases\n",
    "                cdata = sample.data\n",
    "                row = [sample.sample, rec.REF, sample.gt_bases, mut, cdata[2], cdata[4] ,cdata[5], cdata[6]] + x\n",
    "            else:\n",
    "                mut = str(rec.end)+rec.REF\n",
    "                #inf = sample.site.INFO\n",
    "                cdata = sample.data\n",
    "                row = [sample.sample, rec.REF, sample.gt_bases, mut, cdata[2], cdata[4] ,cdata[5], cdata[6]] + x\n",
    "\n",
    "            res.append(row)\n",
    "    res = pd.DataFrame(res,columns=cols)\n",
    "    res = res[~res.start.isnull()]\n",
    "    #res['start'] = res.start.astype(int)\n",
    "    #res['end'] = res.end.astype(int)\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test filters\n",
    "\n",
    "* https://samtools.github.io/bcftools/howtos/filtering.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>mut</th>\n",
       "      <th>DP</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADR</th>\n",
       "      <th>AD</th>\n",
       "      <th>chrom</th>\n",
       "      <th>var_type</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13-11594</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1057A&gt;G</td>\n",
       "      <td>78</td>\n",
       "      <td>[0, 44]</td>\n",
       "      <td>[0, 33]</td>\n",
       "      <td>[0, 77]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14-MBovis</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1057A&gt;G</td>\n",
       "      <td>114</td>\n",
       "      <td>[1, 67]</td>\n",
       "      <td>[0, 46]</td>\n",
       "      <td>[1, 113]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15-11643</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1057A&gt;G</td>\n",
       "      <td>96</td>\n",
       "      <td>[0, 57]</td>\n",
       "      <td>[0, 39]</td>\n",
       "      <td>[0, 96]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17-11662</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1057A&gt;G</td>\n",
       "      <td>98</td>\n",
       "      <td>[0, 59]</td>\n",
       "      <td>[0, 39]</td>\n",
       "      <td>[0, 98]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17-MBovis</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1057A&gt;G</td>\n",
       "      <td>64</td>\n",
       "      <td>[0, 34]</td>\n",
       "      <td>[0, 30]</td>\n",
       "      <td>[0, 64]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16325</th>\n",
       "      <td>55-4348</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4346083G&gt;A</td>\n",
       "      <td>101</td>\n",
       "      <td>[0, 66]</td>\n",
       "      <td>[0, 35]</td>\n",
       "      <td>[0, 101]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>4346082.0</td>\n",
       "      <td>4346083.0</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16326</th>\n",
       "      <td>56-MBovis</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>4346083G</td>\n",
       "      <td>121</td>\n",
       "      <td>[63, 0]</td>\n",
       "      <td>[58, 0]</td>\n",
       "      <td>[121, 0]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>4346082.0</td>\n",
       "      <td>4346083.0</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16327</th>\n",
       "      <td>59-6110</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>4346083G</td>\n",
       "      <td>98</td>\n",
       "      <td>[50, 0]</td>\n",
       "      <td>[44, 0]</td>\n",
       "      <td>[94, 0]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>4346082.0</td>\n",
       "      <td>4346083.0</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16328</th>\n",
       "      <td>59-MBovis</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>4346083G</td>\n",
       "      <td>134</td>\n",
       "      <td>[73, 0]</td>\n",
       "      <td>[61, 0]</td>\n",
       "      <td>[134, 0]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>4346082.0</td>\n",
       "      <td>4346083.0</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16329</th>\n",
       "      <td>7-10423</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>4346083G</td>\n",
       "      <td>110</td>\n",
       "      <td>[56, 0]</td>\n",
       "      <td>[54, 0]</td>\n",
       "      <td>[110, 0]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>ts</td>\n",
       "      <td>4346082.0</td>\n",
       "      <td>4346083.0</td>\n",
       "      <td>211.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14310 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample REF ALT         mut   DP      ADF      ADR        AD       chrom var_type sub_type      start        end   QUAL\n",
       "0       13-11594   A   G     1057A>G   78  [0, 44]  [0, 33]   [0, 77]  LT708304.1      snp       ts     1056.0     1057.0  999.0\n",
       "1      14-MBovis   A   G     1057A>G  114  [1, 67]  [0, 46]  [1, 113]  LT708304.1      snp       ts     1056.0     1057.0  999.0\n",
       "2       15-11643   A   G     1057A>G   96  [0, 57]  [0, 39]   [0, 96]  LT708304.1      snp       ts     1056.0     1057.0  999.0\n",
       "3       17-11662   A   G     1057A>G   98  [0, 59]  [0, 39]   [0, 98]  LT708304.1      snp       ts     1056.0     1057.0  999.0\n",
       "4      17-MBovis   A   G     1057A>G   64  [0, 34]  [0, 30]   [0, 64]  LT708304.1      snp       ts     1056.0     1057.0  999.0\n",
       "...          ...  ..  ..         ...  ...      ...      ...       ...         ...      ...      ...        ...        ...    ...\n",
       "16325    55-4348   G   A  4346083G>A  101  [0, 66]  [0, 35]  [0, 101]  LT708304.1      snp       ts  4346082.0  4346083.0  211.0\n",
       "16326  56-MBovis   G   G    4346083G  121  [63, 0]  [58, 0]  [121, 0]  LT708304.1      snp       ts  4346082.0  4346083.0  211.0\n",
       "16327    59-6110   G   G    4346083G   98  [50, 0]  [44, 0]   [94, 0]  LT708304.1      snp       ts  4346082.0  4346083.0  211.0\n",
       "16328  59-MBovis   G   G    4346083G  134  [73, 0]  [61, 0]  [134, 0]  LT708304.1      snp       ts  4346082.0  4346083.0  211.0\n",
       "16329    7-10423   G   G    4346083G  110  [56, 0]  [54, 0]  [110, 0]  LT708304.1      snp       ts  4346082.0  4346083.0  211.0\n",
       "\n",
       "[14310 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdf = vcf_to_dataframe('test.vcf')\n",
    "vdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print (vdf[vdf.end==2839071])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2799843.0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3673622.0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3673605.0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2799853.0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3488356.0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102980.0</th>\n",
       "      <td>6262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67098.0</th>\n",
       "      <td>6276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2592710.0</th>\n",
       "      <td>6704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193147.0</th>\n",
       "      <td>6773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577163.0</th>\n",
       "      <td>7084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>355 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             DP\n",
       "end            \n",
       "2799843.0     1\n",
       "3673622.0     1\n",
       "3673605.0     1\n",
       "2799853.0     1\n",
       "3488356.0     1\n",
       "...         ...\n",
       "102980.0   6262\n",
       "67098.0    6276\n",
       "2592710.0  6704\n",
       "193147.0   6773\n",
       "3577163.0  7084\n",
       "\n",
       "[355 rows x 1 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdf.groupby('end').agg({'DP':sum}).sort_values(by='DP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "u = vdf.drop_duplicates(['sample','mut']).set_index(['mut','sample'])['ALT'].unstack('sample')\n",
    "u[:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>mut</th>\n",
       "      <th>DP</th>\n",
       "      <th>ADF</th>\n",
       "      <th>ADR</th>\n",
       "      <th>AD</th>\n",
       "      <th>chrom</th>\n",
       "      <th>var_type</th>\n",
       "      <th>sub_type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5290</th>\n",
       "      <td>13-11594</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>11</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>[1, 9]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5291</th>\n",
       "      <td>14-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>12</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5292</th>\n",
       "      <td>15-11643</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>23</td>\n",
       "      <td>[0, 13]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[0, 23]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5293</th>\n",
       "      <td>17-11662</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>6</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[1, 5]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5294</th>\n",
       "      <td>17-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>2</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5295</th>\n",
       "      <td>182-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>2839071C</td>\n",
       "      <td>104</td>\n",
       "      <td>[67, 0]</td>\n",
       "      <td>[37, 0]</td>\n",
       "      <td>[104, 0]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5296</th>\n",
       "      <td>19-11957</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>8</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5297</th>\n",
       "      <td>19-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>4</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5298</th>\n",
       "      <td>22-12200</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5299</th>\n",
       "      <td>23-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>11</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5300</th>\n",
       "      <td>24-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>37</td>\n",
       "      <td>[0, 21]</td>\n",
       "      <td>[0, 15]</td>\n",
       "      <td>[0, 36]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5301</th>\n",
       "      <td>25-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>13</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 13]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5302</th>\n",
       "      <td>26-12883</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>9</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5303</th>\n",
       "      <td>26-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>4</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5304</th>\n",
       "      <td>27-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>14</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[0, 14]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5305</th>\n",
       "      <td>28-12935</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>6</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5306</th>\n",
       "      <td>29-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>13</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5307</th>\n",
       "      <td>3-10110</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[1, 5]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[1, 9]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5308</th>\n",
       "      <td>30-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>6</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5309</th>\n",
       "      <td>31-12952</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>12</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5310</th>\n",
       "      <td>35-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>7</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311</th>\n",
       "      <td>36-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>11</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5312</th>\n",
       "      <td>37-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>7</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5313</th>\n",
       "      <td>38-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>12</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5314</th>\n",
       "      <td>39-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>3</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5315</th>\n",
       "      <td>41-2165</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5316</th>\n",
       "      <td>41-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>16</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 16]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5317</th>\n",
       "      <td>42-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>6</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5318</th>\n",
       "      <td>43-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>18</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 18]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5319</th>\n",
       "      <td>44-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>15</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 15]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5320</th>\n",
       "      <td>45-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>7</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5321</th>\n",
       "      <td>47-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>17</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 17]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5322</th>\n",
       "      <td>48-2919</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>12</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5323</th>\n",
       "      <td>48-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5324</th>\n",
       "      <td>49-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>12</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 12]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5325</th>\n",
       "      <td>5-10284</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5326</th>\n",
       "      <td>50-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>5</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5327</th>\n",
       "      <td>51-3292</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>16</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[0, 16]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5328</th>\n",
       "      <td>51-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>5</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 2]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5329</th>\n",
       "      <td>52-3698</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>11</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5330</th>\n",
       "      <td>54-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>14</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[0, 14]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5331</th>\n",
       "      <td>55-4348</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>9</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5332</th>\n",
       "      <td>56-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>4</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5333</th>\n",
       "      <td>59-6110</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>18</td>\n",
       "      <td>[0, 8]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[0, 17]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5334</th>\n",
       "      <td>59-MBovis</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>10</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5335</th>\n",
       "      <td>7-10423</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2839071C&gt;G</td>\n",
       "      <td>6</td>\n",
       "      <td>[0, 5]</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>[0, 6]</td>\n",
       "      <td>LT708304.1</td>\n",
       "      <td>snp</td>\n",
       "      <td>tv</td>\n",
       "      <td>2839070.0</td>\n",
       "      <td>2839071.0</td>\n",
       "      <td>999.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample REF ALT         mut   DP      ADF      ADR        AD       chrom var_type sub_type      start        end   QUAL\n",
       "5290    13-11594   C   G  2839071C>G   11   [0, 8]   [1, 1]    [1, 9]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5291   14-MBovis   C   G  2839071C>G   12   [0, 9]   [0, 3]   [0, 12]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5292    15-11643   C   G  2839071C>G   23  [0, 13]  [0, 10]   [0, 23]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5293    17-11662   C   G  2839071C>G    6   [1, 1]   [0, 4]    [1, 5]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5294   17-MBovis   C   G  2839071C>G    2   [0, 2]   [0, 0]    [0, 2]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5295  182-MBovis   C   C    2839071C  104  [67, 0]  [37, 0]  [104, 0]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5296    19-11957   C   G  2839071C>G    8   [0, 6]   [0, 2]    [0, 8]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5297   19-MBovis   C   G  2839071C>G    4   [0, 3]   [0, 0]    [0, 3]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5298    22-12200   C   G  2839071C>G   10   [0, 3]   [0, 6]    [0, 9]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5299   23-MBovis   C   G  2839071C>G   11   [0, 7]   [0, 4]   [0, 11]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5300   24-MBovis   C   G  2839071C>G   37  [0, 21]  [0, 15]   [0, 36]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5301   25-MBovis   C   G  2839071C>G   13  [0, 10]   [0, 3]   [0, 13]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5302    26-12883   C   G  2839071C>G    9   [0, 7]   [0, 2]    [0, 9]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5303   26-MBovis   C   G  2839071C>G    4   [0, 3]   [0, 1]    [0, 4]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5304   27-MBovis   C   G  2839071C>G   14   [0, 6]   [0, 8]   [0, 14]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5305    28-12935   C   G  2839071C>G    6   [0, 3]   [0, 3]    [0, 6]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5306   29-MBovis   C   G  2839071C>G   13   [0, 7]   [0, 5]   [0, 12]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5307     3-10110   C   G  2839071C>G   10   [1, 5]   [0, 4]    [1, 9]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5308   30-MBovis   C   G  2839071C>G    6   [0, 3]   [0, 3]    [0, 6]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5309    31-12952   C   G  2839071C>G   12   [0, 4]   [0, 7]   [0, 11]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5310   35-MBovis   C   G  2839071C>G    7   [0, 6]   [0, 1]    [0, 7]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5311   36-MBovis   C   G  2839071C>G   11   [0, 9]   [0, 2]   [0, 11]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5312   37-MBovis   C   G  2839071C>G    7   [0, 5]   [0, 2]    [0, 7]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5313   38-MBovis   C   G  2839071C>G   12   [0, 4]   [0, 8]   [0, 12]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5314   39-MBovis   C   G  2839071C>G    3   [0, 1]   [0, 2]    [0, 3]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5315     41-2165   C   G  2839071C>G   10   [0, 4]   [0, 6]   [0, 10]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5316   41-MBovis   C   G  2839071C>G   16  [0, 11]   [0, 5]   [0, 16]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5317   42-MBovis   C   G  2839071C>G    6   [0, 4]   [0, 2]    [0, 6]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5318   43-MBovis   C   G  2839071C>G   18  [0, 11]   [0, 7]   [0, 18]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5319   44-MBovis   C   G  2839071C>G   15  [0, 12]   [0, 3]   [0, 15]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5320   45-MBovis   C   G  2839071C>G    7   [0, 6]   [0, 1]    [0, 7]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5321   47-MBovis   C   G  2839071C>G   17  [0, 11]   [0, 6]   [0, 17]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5322     48-2919   C   G  2839071C>G   12   [0, 7]   [0, 5]   [0, 12]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5323   48-MBovis   C   G  2839071C>G   10   [0, 6]   [0, 4]   [0, 10]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5324   49-MBovis   C   G  2839071C>G   12   [0, 8]   [0, 4]   [0, 12]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5325     5-10284   C   G  2839071C>G   10   [0, 5]   [0, 5]   [0, 10]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5326   50-MBovis   C   G  2839071C>G    5   [0, 4]   [0, 1]    [0, 5]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5327     51-3292   C   G  2839071C>G   16   [0, 7]   [0, 9]   [0, 16]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5328   51-MBovis   C   G  2839071C>G    5   [0, 3]   [0, 2]    [0, 5]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5329     52-3698   C   G  2839071C>G   11   [0, 5]   [0, 6]   [0, 11]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5330   54-MBovis   C   G  2839071C>G   14   [0, 6]   [0, 8]   [0, 14]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5331     55-4348   C   G  2839071C>G    9   [0, 6]   [0, 3]    [0, 9]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5332   56-MBovis   C   G  2839071C>G    4   [0, 3]   [0, 1]    [0, 4]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5333     59-6110   C   G  2839071C>G   18   [0, 8]   [0, 9]   [0, 17]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5334   59-MBovis   C   G  2839071C>G   10   [0, 5]   [0, 5]   [0, 10]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0\n",
       "5335     7-10423   C   G  2839071C>G    6   [0, 5]   [0, 1]    [0, 6]  LT708304.1      snp       tv  2839070.0  2839071.0  999.0"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd = 'bcftools filter -e \"QUAL<40 & FMT/DP<30\" -o test.vcf.gz -O z test.vcf'\n",
    "subprocess.check_output(cmd,shell=True)\n",
    "fdf = vcf_to_dataframe('test.vcf.gz')\n",
    "fdf[fdf.end==2839071]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make fasta alignment from filtered VCF sites with multiple samples\n",
    "\n",
    "https://github.com/JosephCrispell/GeneralTools/blob/master/ProcessingPipeline/CreateFastaWithReferenceFromFiltered_28-06-17.pl\n",
    "\n",
    "\"The allele frequencies at each position in the aligned (against reference) sequence from each isolate were examined. For a haploid organism these frequencies are expected to be either 0 or 1, with some random variation expected from sequencing errors (Sobkowiak et al., 2018). A heterozygous site was defined as one where the allele frequencies were >0.05 and <0.95. Four cattle-derived sequences that had more than 150 heterozygous sites, and allele frequencies that were clustered and non-random (data not shown), were removed. In addition, 26 badger-derived and 16 cattle-derived M. bovis sequences were removed because of suspected errors in the metadata (Appendix 1: Investigating isolate metadata discrepancies).\n",
    "\n",
    "For the sequences from the remaining isolates (204 badger- and 169 cattle-derived isolates), alleles were called at each variant position if they had mapping quality ≥30, high-quality base depth ≥4 (applied to reverse and forward reads separately), read depth ≥30, and allele support ≥0.95. For any site that failed these criteria, if the allele called had been observed in a different isolate that had passed, a second round of filtering was conducted using a high-quality base depth of 5 (total across forward and reverse reads) and the same allele support. As recombination is thought to be extremely rare for mycobacteria (Namouchi et al., 2012), variants in close proximity could indicate a region that is difficult to sequence or under high selection. To avoid calling variants in these regions, variant positions within 10 bp of one another were removed. Following filtering, sequences from 11 badger and 10 cattle isolates that had insufficient coverage (<95%) of the variant positions were removed. Once the alignment was generated, sites with a consistency index less than 1, generally considered homoplasies (Farris, 1989), were removed (n = 4, of 14,991 sites) using HomoplasyFinder (v0.0.0.9; Crispell et al., 2019; RRID: SCR_017300). All the scripts necessary for the processing of the WGS data are freely available online.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 547 sites\n",
      "23 sites with at least one missing sample\n",
      "191 uninformative sites\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "48"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def fasta_alignment_from_vcf(vcf_file, callback=None, uninformative=False, omit=None):\n",
    "    \"\"\"\n",
    "    Get core SNP site calls as sequences from a multi sample vcf file.\n",
    "    Args:\n",
    "        vcf_file: multi-sample vcf (e.g. produced by app.variant_calling)\n",
    "        uninformative: whether to include uninformative sites\n",
    "        omit: list of samples to exclude if required\n",
    "    \"\"\"\n",
    "\n",
    "    import vcf\n",
    "    from collections import defaultdict\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    #print (vcf_reader.samples)\n",
    "    def default():\n",
    "        return []\n",
    "    result = defaultdict(default)\n",
    "    sites = []\n",
    "    result['ref'] = []\n",
    "    missing = []\n",
    "    unf = []\n",
    "    for record in vcf_reader:\n",
    "        S = {sample.sample: sample.gt_bases for sample in record.samples}\n",
    "        if omit != None:\n",
    "            for o in omit:\n",
    "                del S[o]\n",
    "        #if any missing samples at the site we don't add\n",
    "        if None in S.values():\n",
    "            #for i in S:\n",
    "            #    if S[i]==None:\n",
    "            #        print (i)\n",
    "            missing.append(record.POS)\n",
    "            continue\n",
    "        if uninformative == False:\n",
    "            u = set(S.values())\n",
    "            if len(u) == 1:\n",
    "                unf.append(record.POS)\n",
    "                continue\n",
    "        result['ref'].append(record.REF)\n",
    "        for name in S:\n",
    "            result[name].append(S[name])\n",
    "        sites.append(record.POS)\n",
    "\n",
    "    print ('found %s sites' %len(sites))\n",
    "    print ('%s sites with at least one missing sample' %len(missing))\n",
    "    if uninformative == False:\n",
    "        print ('%s uninformative sites' %len(unf))\n",
    "    if len(sites)==0:\n",
    "        print ('no sites found may mean one sample is too different')\n",
    "    recs = []\n",
    "    for sample in result:\n",
    "        seq = ''.join(result[sample])\n",
    "        seqrec = SeqRecord(Seq(seq),id=sample)\n",
    "        recs.append(seqrec)\n",
    "        #print (len(seqrec))\n",
    "\n",
    "    smat = pd.DataFrame(result)\n",
    "    smat.index = sites\n",
    "    smat.index.rename('pos', inplace=True)\n",
    "    return recs, smat\n",
    "\n",
    "#seqrecs, smat = fasta_alignment_from_vcf('mapped/filtered.vcf.gz')\n",
    "seqrecs, smat = fasta_alignment_from_vcf('/storage/btbgenie/wicklow_results/filtered.vcf.gz')\n",
    "SeqIO.write(seqrecs, 'core.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "proximity filter found 11/629 sites\n"
     ]
    }
   ],
   "source": [
    "def site_proximity_filter(vcf_file, dist=10, overwrite=False, outdir=None):\n",
    "    \"\"\"Remove any pairs of sites within dist of each other\"\"\"\n",
    "\n",
    "    import vcf\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    sites = [record.POS for record in vcf_reader]\n",
    "    #print (sites)\n",
    "    found = []\n",
    "    for i in range(len(sites)-1):\n",
    "        if sites[i+1] - sites[i] <= dist:\n",
    "            #print (sites[i], sites[i+1],'close')\n",
    "            found.extend([sites[i], sites[i+1]])\n",
    "    #print (found)\n",
    "    new = sorted(list(set(sites) - set(found)))\n",
    "    print ('proximity filter found %s/%s sites' %(len(set(found)),len(sites)))\n",
    "    if outdir == None:\n",
    "        outdir = tempfile.gettempdir()\n",
    "    if overwrite == True:\n",
    "        out = os.path.join(outdir,'temp.vcf')\n",
    "        vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "        vcf_writer = vcf.Writer(open(out, 'w'), vcf_reader)\n",
    "        for record in vcf_reader:\n",
    "            if record.POS in new:\n",
    "                #print (record)\n",
    "                vcf_writer.write_record(record)\n",
    "        vcf_writer.close()\n",
    "        #overwrite input vcf\n",
    "        bcftoolscmd = tools.get_cmd('bcftools')\n",
    "        cmd = 'bcftools view {o} -O z -o {gz}'.format(o=out,gz=vcf_file)\n",
    "        tmp = subprocess.check_output(cmd,shell=True)\n",
    "    return new\n",
    "\n",
    "old = site_proximity_filter('/storage/btbgenie/sim_results/snps.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       sample REF ALT      mut   DP       ADF      ADR        AD       chrom var_type sub_type   pos   start     end   QUAL\n",
      "0    13-11594   A   G  1057A>G  170  [0, 125]  [0, 45]  [0, 170]  LT708304.1      snp       ts  1057  1056.0  1057.0  999.0\n",
      "1   14-MBovis   A   G  1057A>G  188  [0, 158]  [0, 30]  [0, 188]  LT708304.1      snp       ts  1057  1056.0  1057.0  999.0\n",
      "2    15-11643   A   G  1057A>G  165  [0, 158]   [0, 7]  [0, 165]  LT708304.1      snp       ts  1057  1056.0  1057.0  999.0\n",
      "3  161-MBovis   A   G  1057A>G  136  [0, 135]   [0, 1]  [0, 136]  LT708304.1      snp       ts  1057  1056.0  1057.0  999.0\n"
     ]
    }
   ],
   "source": [
    "df = tools.vcf_to_dataframe('/storage/btbgenie/sim_results/snps.vcf.gz') \n",
    "print (df[:4])\n",
    "p=pd.pivot_table(df,index='pos',columns='sample',values='mut',aggfunc='size')\n",
    "#p[p.isnull().sum(1)>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = [2672966,2916881,3499480,3499469,3499476,2672970,2444846]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "diff = list (set(new) - set(old))\n",
    "sub=df[df.pos.isin(diff)]\n",
    "p=pd.pivot_table(sub,index='pos',columns='sample',values='mut',aggfunc='first')\n",
    "#import seaborn as sns\n",
    "#sns.heatmap(p)\n",
    "p.to_csv('proximity_changed.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overwrite_vcf(vcf_file, sites, outdir=None):\n",
    "    \"\"\"Make a new vcf with subset of sites\"\"\"\n",
    "    \n",
    "    if outdir == None:\n",
    "        outdir = tempfile.gettempdir()\n",
    "    \n",
    "    out = os.path.join(outdir,'temp.vcf')\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    vcf_writer = vcf.Writer(open(out, 'w'), vcf_reader)\n",
    "    for record in vcf_reader:\n",
    "        if record.POS in sites:\n",
    "            #print (record)\n",
    "            vcf_writer.write_record(record)\n",
    "    vcf_writer.close()\n",
    "    #copy or overwrite input vcf\n",
    "    bcftoolscmd = tools.get_cmd('bcftools')\n",
    "    cmd = 'bcftools view {o} -O z -o {gz}'.format(o=out,gz=vcf_file)\n",
    "    tmp = subprocess.check_output(cmd,shell=True)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "proximity filter found 0/629 sites\n"
     ]
    }
   ],
   "source": [
    "reload(tools)\n",
    "\n",
    "def new_site_proximity_filter(vcf_file, dist=10, overwrite=False, outdir=None):\n",
    "    \"\"\"Remove any pairs of sites within dist of each other.\n",
    "    Args:\n",
    "        vcf_file: input vcf file with positions to filter\n",
    "        dist: distance threshold\n",
    "        overwrite: whether to overwrite the vcf\n",
    "    \"\"\"\n",
    "\n",
    "    #get vcf into dataframe\n",
    "    df = tools.vcf_to_dataframe(vcf_file)\n",
    "    df = df[df.REF != df.ALT]\n",
    "    sites = list(df.pos.unique())\n",
    "    found = []\n",
    "    #check distances in sites per sample\n",
    "    for s, g in df.groupby(['sample']):\n",
    "        #print (s)\n",
    "        pos = list(g.pos)\n",
    "        a = list(g.mut)   \n",
    "        for i in range(len(pos)-1):\n",
    "            #if pos[i] in fn:\n",
    "                #print (s,pos[i],pos[i+1], pos[i+1] - pos[i],alt[i])\n",
    "                #print (g[g.pos==pos[i]])            \n",
    "            if pos[i+1] - pos[i] <= dist:\n",
    "                found.extend([pos[i], pos[i+1]])\n",
    "    #all unique positions    \n",
    "    found = list(set(found))\n",
    "    print (set(found).intersection(set(fn)))\n",
    "    new = sorted(list(set(sites) - set(found)))\n",
    "    print ('proximity filter found %s/%s sites' %(len(found),len(sites)))\n",
    "    if overwrite == True:    \n",
    "        overwrite_vcf(vcf_file, new, outdir)\n",
    "    return new\n",
    "\n",
    "new=new_site_proximity_filter('/storage/btbgenie/sim_results/snps.vcf.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mask locations for mbovis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make mask bed file for mbovis\n",
    "g=tools.genbank_to_dataframe('../snipgenie/data/Mbovis_AF212297.gb')\n",
    "g['gene'] = g.gene.fillna('')\n",
    "g=g[(g.gene.str.contains('PE_PGRS') | g.gene.str.contains('PPE')) & (g.feat_type=='CDS') | (g.feat_type=='repeat_region')]\n",
    "for i,r in g.iterrows():  \n",
    "    print ('LT708304.1','\\t', r.start, '\\t', r.end, r.gene, r.locus_tag)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mask for MTB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make mask bed file for mbovis\n",
    "g=tools.genbank_to_dataframe('../snipgenie/data/MTB-H37Rv.gb')\n",
    "g['gene'] = g.gene.fillna('')\n",
    "g=g[(g.gene.str.contains('PE_PGRS') | g.gene.str.contains('PPE')) & (g.feat_type=='CDS') | (g.feat_type=='repeat_region')]\n",
    "for i,r in g.iterrows():  \n",
    "    print ('NC_000962.3','\\t', r.start, '\\t', r.end)#, r.gene, r.locus_tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mask_filter(vcf_file, mask_file):\n",
    "    \"\"\"Remove any masked sites using a bed file, overwrites input\"\"\"\n",
    "\n",
    "    print('using mask bed file', mask_file)\n",
    "    mask = pd.read_csv(mask_file,sep='\\t',names=['chrom','start','end'])    \n",
    "    #print (mask)\n",
    "    def do_mask(x,i):     \n",
    "        #print (x.start, x.end, i)\n",
    "        if (x.start<=i) & (x.end>=i):            \n",
    "            return 1\n",
    "    import vcf\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    sites = [record.POS for record in vcf_reader]\n",
    "    found = []\n",
    "    for i in sites:       \n",
    "        m = mask.apply( lambda x: do_mask(x,i),1)        \n",
    "        m = m[m==1]   \n",
    "        if len(m)>0:    \n",
    "            print (i)\n",
    "            found.append(i)\n",
    "    print('found %s sites in masked regions' %len(found))\n",
    "    new = sorted(list(set(sites) - set(found)))\n",
    "    \n",
    "    tempdir = tempfile.gettempdir()\n",
    "    out = os.path.join(tempdir,'temp.vcf')\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    vcf_writer = vcf.Writer(open(out, 'w'), vcf_reader)\n",
    "    for record in vcf_reader:\n",
    "        if record.POS in new:\n",
    "            #print (record)\n",
    "            vcf_writer.write_record(record)\n",
    "    vcf_writer.close()\n",
    "    #overwrite input vcf\n",
    "    bcftoolscmd = tools.get_cmd('bcftools')\n",
    "    cmd = 'bcftools view {o} -O z -o {gz}'.format(o=out,gz=vcf_file)\n",
    "    tmp = subprocess.check_output(cmd,shell=True)\n",
    "    return\n",
    "\n",
    "mask_filter('mapped/filtered.vcf.gz','../snpgenie/data/Mbovis_AF212297_mask.bed')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## uninformative sites\n",
    "\n",
    "probably not needed as we can use `--min-ac 1:minor`  in bcftools view?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "629 78\n"
     ]
    }
   ],
   "source": [
    "def uninformative_sites_filter(vcf_file):\n",
    "    \"\"\"Ree\"\"\"\n",
    "\n",
    "    import vcf\n",
    "    vcf_reader = vcf.Reader(open(vcf_file, 'rb'))\n",
    "    i=0; u=0\n",
    "    for record in vcf_reader:\n",
    "        ref=record.REF\n",
    "        S = set([sample.gt_bases for sample in record.samples])\n",
    "        #print (record.POS,ref,S)\n",
    "        if (len(S)==1):\n",
    "            #print (record.POS,ref,S)\n",
    "            u+=1\n",
    "        i+=1\n",
    "    print (i,u)    \n",
    "    \n",
    "#u_filter('mapped/filtered.vcf.gz')\n",
    "#u_filter('../test_results/filtered.vcf.gz')\n",
    "uninformative_sites_filter('/storage/btbgenie/wicklow_results/filtered.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generic_multiprocess(inputs):\n",
    "    #split sites into chunks and run in parallel for long jobs\n",
    "    \n",
    "    import multiprocessing as mp\n",
    "    blocks = np.array_split(np.array(inputs),threads)\n",
    "    result=[]\n",
    "    funclist = []\n",
    "    pool = mp.Pool(threads)\n",
    "    for subset in blocks:            \n",
    "        f = pool.apply_async(worker_func, [subset])      \n",
    "        funclist.append(f)            \n",
    "    try:\n",
    "        for f in funclist:\n",
    "            out = f.get(timeout=None)               \n",
    "            if len(out)>0:\n",
    "                result.append(out)\n",
    "    except KeyboardInterrupt:\n",
    "        print ('process interrupted')\n",
    "        pool.terminate()\n",
    "        sys.exit(0)\n",
    "\n",
    "    pool.close()\n",
    "    pool.join()    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogeny from MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_RAXML(infile, name='variants', threads=8):\n",
    "    \"\"\"Run Raxml pthreads\"\"\"\n",
    "\n",
    "    bootstraps = 10\n",
    "    model = 'GTRCAT'\n",
    "    s1 = random.randint(0,1e8)\n",
    "    s2 = random.randint(0,1e8)\n",
    "    \n",
    "    files = glob.glob('RAxML_*')\n",
    "    for f in files:\n",
    "        os.remove(f)\n",
    "    cmd = 'raxmlHPC-PTHREADS -f a -N {nb} -T {t} -m {m} -V -p {s1} -x {s2} -n {n} -s {i}'.format(t=threads,nb=bootstraps,n=name,i=infile,s1=s1,s2=s2,m=model)\n",
    "    print (cmd)\n",
    "    tmp = subprocess.check_output(cmd, shell=True)\n",
    "    return\n",
    "    \n",
    "run_RAXML('variants.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_tree(filename, labelmap=None):\n",
    "    \"\"\"Draw a tree \"\"\"\n",
    "    \n",
    "    from ete3 import Tree, NodeStyle, TreeStyle\n",
    "    t = Tree(filename)\n",
    "    t.set_outgroup('ref')    \n",
    "    if labelmap != None:\n",
    "        trees.set_tiplabels(t,labelmap)\n",
    "    trees.format_nodes(t)\n",
    "    ts = TreeStyle()\n",
    "    ts.scale=1000\n",
    "    t.render(\"%%inline\", tree_style=ts)\n",
    "    t.render(\"tree.png\", tree_style=ts)\n",
    "    return t\n",
    "\n",
    "sra['filename'] = sra.Run.apply(lambda x: 'mapped/'+x+'.bam')\n",
    "print (sra[:3])\n",
    "labelmap = dict(zip(sra.filename,sra.geo_loc_name_country))\n",
    "#print (labelmap)\n",
    "t = create_tree('RAxML_bipartitions.variants', labelmap)\n",
    "t.render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "tree=Phylo.read('RAxML_bestTree.variants','newick')\n",
    "Phylo.draw(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## csq gff format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tools.gff_bcftools_format('../snpgenie/data/Mbovis_AF212297.gb', '../Mbovis_csq_new.gff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff=tools.gff_to_features('../pathogenie/data/Mbovis_AF212297.gff')\n",
    "g=tools.features_to_dataframe(gff)\n",
    "tools.features_summary(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(plotting)\n",
    "f,ax=plt.subplots(1,1,figsize=(12,3))\n",
    "st=time.time()\n",
    "plotting.plot_bam_alignment('../test_results/mapped/3_S57_L001.bam', 'LT708304.1', 2000, 3000, yend=100, rect_height=3, ax=ax)\n",
    "print (time.time()-st)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_features(rec, ax, rows=3, xstart=0, xend=30000):\n",
    "    \n",
    "    h=1\n",
    "    df = tools.records_to_dataframe([rec])    \n",
    "    df = df[(df.feat_type!='region') & (df['feat_type']!='source')]\n",
    "    df = df[(df.start>xstart) & (df.end<xend)]\n",
    "    df['length'] = df.end-df.start\n",
    "    y = list(range(1,rows)) * len(df)\n",
    "    df['y'] = y[:len(df)]\n",
    "    df['color'] = 'blue'\n",
    "    df = df.fillna('')\n",
    "    #print (df)  \n",
    "    \n",
    "    from matplotlib.collections import PatchCollection\n",
    "    import matplotlib.patches as mpatches\n",
    "\n",
    "    patches=[]\n",
    "    for i,r in df.iterrows():        \n",
    "        #rect = plt.Rectangle((r.start, r.y), r.length, h/2,\n",
    "        #                        alpha=.5, linewidth=.5,\n",
    "        #                        edgecolor='black', facecolor=r.color)\n",
    "        if r.strand == 1:\n",
    "            x = r.start\n",
    "            dx = r.length\n",
    "        else:\n",
    "            x = r.end\n",
    "            dx = -r.length\n",
    "        arrow = mpatches.Arrow(x, r.y, dx, 0, alpha=.7, width=.3, \n",
    "                               edgecolor='black')\n",
    "        txt = ax.text(r.start, r.y-h/2, r.gene, size=16)\n",
    "        patches.append(arrow)  \n",
    "        \n",
    "    ax.add_collection(PatchCollection(patches, match_original=True))  \n",
    "    ax.set_xlim(xstart, xend)\n",
    "    ax.set_ylim(.4,rows-.5)\n",
    "    plt.yticks([])\n",
    "    plt.tight_layout()\n",
    "\n",
    "    def onclick(event):\n",
    "        print('%s click: button=%d, x=%d, y=%d, xdata=%f, ydata=%f' %\n",
    "              ('double' if event.dblclick else 'single', event.button,\n",
    "               event.x, event.y, event.xdata, event.ydata))\n",
    "        ax.text(event.x, event.y, 'HI!')\n",
    "        ax.figure.canvas.draw()\n",
    "    #cid = ax.figure.canvas.mpl_connect('button_press_event', onclick)\n",
    "    return\n",
    "\n",
    "reload(tools)\n",
    "recs = tools.gff_to_records(app.mbovis_gff)\n",
    "rec=recs[0]\n",
    "f,ax=plt.subplots(1,1,figsize=(15,1.5))\n",
    "plot_features(rec, ax=ax, xstart=26000,xend=35000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## replace pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chrom(bam_file):\n",
    "    \"\"\"Get first sequence name in a bam file\"\"\"\n",
    "\n",
    "    import pysam\n",
    "    samfile = pysam.AlignmentFile(bam_file, \"r\")\n",
    "    iter=samfile.fetch(start=0,end=10)\n",
    "    for read in iter:\n",
    "        if read.reference_name:\n",
    "            return read.reference_name\n",
    "        \n",
    "get_chrom('mapped/3_S57_L001.bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bamnostic as bs\n",
    "bam = bs.AlignmentFile('mapped/3_S57_L001.bam', 'rb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LT708304.1'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom=bam.header['SQ'][0]['SN']\n",
    "chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, read in enumerate(bam.fetch(chrom, 1, 10)):\n",
    "    print(read)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coverage(bam_file, chr, start, end):\n",
    "    \"\"\"Get coverage from bam file at specified region\"\"\"\n",
    "    \n",
    "    import bamnostic as bs\n",
    "    if bam_file is None or not os.path.exists(bam_file):\n",
    "        return\n",
    "    bam = bs.AlignmentFile(bam_file, 'rb')\n",
    "    for i, read in enumerate(bam.fetch(chrom, 1, 10)):\n",
    "        print(read)\n",
    "    return\n",
    "\n",
    "get_coverage('mapped/3_S57_L001.bam', chrom, 1, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def samtools_flagstats(filename):\n",
    "    \"\"\"Parse samtools flagstat output into dictionary\"\"\"\n",
    "    \n",
    "    cmd = 'samtools flagstat %s' %filename\n",
    "    tmp = subprocess.check_output(cmd, shell=True, universal_newlines=True)\n",
    "    x = tmp.split('\\n')    \n",
    "    x = [int(i.split('+')[0]) for i in x[:-1]]\n",
    "    #print (x)\n",
    "    cols = ['total','secondary','supplementary','duplicates','mapped',\n",
    "            'paired','read1','read2','properly paired','with itself','singletons']\n",
    "    d = {}\n",
    "    for c,v in zip(cols,x):\n",
    "        d[c] = v\n",
    "    return d\n",
    "\n",
    "samtools_flagstats('mapped/6_S7_L001.bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def samtools_tview(bam_file, chrom, pos, ref='', display='T'):\n",
    "    \"\"\"View bam alignment with samtools\"\"\"\n",
    "    \n",
    "    cmd = 'COLUMNS=150 samtools tview {b} -p {c}:{p} -d {d} {r}'.format(b=bam_file,c=chrom,p=pos,d=display,r=ref)\n",
    "    #print (cmd)\n",
    "    tmp = subprocess.check_output(cmd, shell=True, universal_newlines=True)    \n",
    "    return tmp\n",
    "\n",
    "t=samtools_tview('mapped/6_S7_L001.bam', 'LT708304.1', 1900, app.mbovis_genome)\n",
    "print (t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def samtools_depth(bam_file, chrom, start, end):\n",
    "    \"\"\"Get depth from bam file\"\"\"\n",
    "\n",
    "    cmd = 'samtools depth -r {c}:{s}-{e} {b}'.format(b=bam_file,c=chrom,s=start,e=end)\n",
    "    tmp=subprocess.check_output(cmd, shell=True, universal_newlines=True)\n",
    "    from io import StringIO\n",
    "    c = pd.read_csv(StringIO(tmp),sep='\\t',names=['chr','pos','depth'])\n",
    "    return c\n",
    "    \n",
    "t=samtools_depth('mapped/6_S7_L001.bam', 'LT708304.1', 1900, 2000)\n",
    "t"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/storage/btbgenie/mbovis_ireland/Wicklow/Fastqs_15-03-19/'\n",
    "\n",
    "#def get_subsample_reads(path, reads):\n",
    "df=app.get_samples(app.get_files_from_paths(path),'_')\n",
    "df=df.sort_values('sample')\n",
    "names = df['sample'].unique()[:3]\n",
    "reads=400000\n",
    "lines = reads*4\n",
    "print (lines % 4)\n",
    "for i,r in list(df.iterrows())[:8]:\n",
    "    f = r.filename  \n",
    "    cmd = 'zcat {f} | head -n {r} | gzip > ../testing/{n}.fastq.gz'.format(f=f,r=lines,n=r['name'])\n",
    "    print (cmd)\n",
    "    subprocess.check_output(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "v1 = tools.vcf_to_dataframe('../test_results/filtered.vcf.gz')\n",
    "v2 = tools.vcf_to_dataframe('../subread_results/filtered.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib_venn._common.VennDiagram at 0x7f369e458820>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAO8AAADsCAYAAACc5PEXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAd9ElEQVR4nO2deXRc9ZXnP1elqtJqrTaSjfEC2BgCNgZDHBzHAUKAsKRDJ910h06Tw2RIJ8xM0jlZGjJFnQBJH5KZbqaZTtJ9OtuESQiQhDBhidkXgxMWG7AN3ndhy9pVUq2/+eOVTCGXbElW1Xvv9+7nnHekeq+k362SvnV/7/7uvT8xxqAoiv+ocNsARVEmh4pXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8iopXUXyKildRfIqKV1F8SqXbBijHh8QlBNQUOSI4f99Q/pCRHwEMkCo4kgVfk8AAMGBi2hfYy4j2bfYHEpcaoBloyX9tBuqAaImGzAF9+aO34OshEzNDJRpTmQAqXg8icYkA7fmjFUeoVa4a9V76gHcKji710uVHxesB8lPfE4BZ+aMVf8Uj0sBBYB+wy8RMp8v2BAIVr0tIXKqAefmjDbviD4PALmA7sM/ETM5le6xExVtG8tPhecApwEzeDSLZTBLYCWwxMbPHbWNsQsVbBiQus4EFwBzs8rATpR/YBLxlYibhtjF+R8VbIiQulTiCfR/Q6K41niOHM63eoN548qh4pxiJSy1wBrCI0i3j2EQ/sB7YZGIm67YxfkLFO0VIXJqBJcB8/BUp9goJ4DUcEWdctsUXqHiPE4lLPXAuThAqCAGoUpPA8cQbVMRHR8U7SfJLPUuB01FPWwqGcDzxm7rUVBwV7wTJB6LOyh8Rl80JAt3A8yZm9rltiNdQ8U4Aictc4AKg1mVTgshW4EUTM4NuG+IVVLzjQOJSjSPa+W7bEnDSwCvA6zqVVvEeE4nLAmA5uuzjJXqAp0zMHHDbEDdR8Y6BxKUOWAmc6LYtSlFyOAGtV4LqhVW8RZC4nAqsAMJu26Ick4PA4yZm+tw2pNyoeAvIR5IvABa6bYsyIdLAcyZmNrttSDlR8eaRuDQAH8EpfFf8ydvAs0FJs1TxcngJaBW6bmsDB4DHglC1FHjxSlzOxcmUUuxhAHjUxMwhtw0pJYEVr8SlAieavMBtW5SSkAGeMDGzw21DSkUgxZsPTH0EmD1Vv7MuS3Z2kkx7mtzMFKYtDdPTSJVBBEQMBpwkaMkXMFQUvPU9leT2RjB7IrAnguyJUHkgTGVWtNjhODDAWhMz69w2pBQETrz5goJLgRmT/R2zkqTOSpA5JYmZm6RiZopwXW7qO2RkwfSFyBwMk9kfxuyMwto6wjujuoQ1QTaYmHnObSOmmkCJN1++dznQMKGfM5izEiRX9pNdOkikNeOueHpCpNfVkH6+HnmllmiyQquaxsEmnEi0Nf/wgRGvxKURuAJnN4FjUpMlu3yA5Ip+eF+CaJUhVFIDJ0kGctuqSL5Yh3munvD+iHrlo/A28LQtAg6EeCUu04CrOIZwq3LkPtbN8Ip+KuYliYZ8WFzfHSL9bD2pXzcT7QwHutndWGwBnrRBwNaLN99T6iqgfqznNGbIXHuI5EW9VEU96mEnShbM+hqG7m+mYl2tp3Zb8AJbcQTs65xoq8WbD05dxRjdG9tTpK/rJL28n6pKi7thvFNJ6pctZFY3UG00ej3CFhMzT7htxPFgrXjzDc6vxNmY6z2cOkTyuk6yixNUV/hwajxZukOkf9VC+veNVOsSFACvmZhZ67YRk8VK8ebXca9g1HLQ2YMMX3cQTk0GexrZGyL98xbSDzeNL3hnOc+amNnothGTwVbxXkxB14u6LNkv7Sd53qD+sxayOcrwnTMJBTxCncNJpdzttiETxTrxSlzOBpaNPP5wL4n/fIBIbQmSKGwgDbn7Wxj6RQs1AZ5Kp4EH/ZYLbZV4JS4nAR8FpDlD5sv7SS9OUO22XX5gX5jkd9thc3Vg2/0MAr/xU4M7a8SbT8L4OBC5rJvEZw96N7HCq+TArG4g8cMZVAc0a+sAjgf2xRKSFX+gfGT5kulp5M6dDP3dAWpUuBOnAuSSXmr/fRvZcwYYdtseF5gBnO+2EePFCvECqz7YR/T72wmdNqzT5OOlMUv4v+8les0hrC9oL8KZEpd5bhsxHnw/bZa4LLruIEs/1aWN0EvB49NI/HNb4JI7ksD9JmYG3DbkaPhavO03y7QbD/CxcwbHTn1Ujp83qxmKnRi46qX9wENezoH27R/jwEyJ/Mt2PqzCLT1nDFF91w7SzRmCtGtfO86WrZ7Fn+IVqd13EpdG4QS3TQkKM9NE79qBmTdMym1bysg5+VUMT+I/8Yo09DRx2Z65Ktxy05AlfOcuQssGGHLbljJRgdN835P4S7wi9cAVry2nlopABVA8Q9QQunkvVR/sC4yAZ+b3q/Ic/hGvSDVw+Y5TkIGGwGYBeYIQyJf2E12UIOm2LWXi/RIXz/3P+UO8ImHgskwl9ZsWa2N0LxCGilv3UNmeIu22LWWgCg8mb3hfvCIhnHzl1o1LGMpEtMDAK9QYQt/ehZmWIQjbi5wmcWlz24hCvC1eEQEuBGYO1pHedbJmT3mNliyR23eTDufwRT7wcfIBtw0oxNvidTa1ngew7jwyJlhJAr5hboqqb+4NRC50a35fK0/gXTGIzAPeB3DwBIa7TlCv62XOTlDzhQ58U053HJwrcfHESoc3xStSB3xo5OGmxS7aooybS3upvbrL+mKGZuBkt40AL4pXpAK4mPx2m12tJHtbgt1zyk/87UGq5iStz8I6xwve13vidVrYHG4ct3Exnk0MV46kEiq+sRcTMlb/3RrwwO6S3hKvyGzg8CS5p4lk9wz1un5jVpro3xy0fvq81G3v6x3xikRwdqc/zKYlgVg/tJKru6k5dcjqDKx6YI6bBnhHvHAuvBtRHqgn1akRZt8SAvnKfmeHRbdtKSFnuDm4N8Qr0gScXnhq6yLSBKt7g3XMTBP9RJfVBQyz3CwZ9IZ4ncyVw7bkBLPvJL3XtYFrDxG1vIjfNe/rvnidZIxZhaf2zWEoG9bujzYQNYRu6rB66ehUiYsrO064K16n6OD9o0/vWKDTZZs4d5CahfYGryLAqW4M7LbnPYtR++YO1pLuadFAlW38dafVKweL3BjUPfE6NbpnjT69faHVU6zAsjhBtcW1vy1uBK7c9LyL4MiOGPtP0mJ7G6kAufaQteIFOKXcA7ojXude9wiv29tIOlkd6O0mrWZFH1UWF+7PP/ZTpha3PO8pcOReufvm6JTZZsJQcU2XtXW/jRKXpnIO6JZ4i66NdZyoLW5s59IeqqL2dt2YW87Byi9ekTagdfTpoWoyg9O0K6Tt1BhCl/dY633nlnMwNzzv6cVO6pQ5OPxZF2FLc56nS1zKtsxZXvE6gaqilRj7Z2tiRlBoyhI+d9Ba79teroHK7Xlnw5HRZAOmr1GXiILEBf1Wel6wWLxFNy3uayKdq9Rc5iCxdNDaJUELxev0pio6Ze48weqqE6UITVnCs+3sddVUrq1Ryul5T4TiU+NO3e8vkKzotzLjSoCy7KxQTvGOmYHS22ztFEo5CucPWHurVJapcznFe1Kxk8PVZFJVKt4gMjdJtCZrZbrkjGM/5fgpj3hFGqB4Z4yuViunTso4CIGcP2BlnW9Z0iTL5XnH/CTqb7Q2VU4ZBx8YcNuCkhAtR7KG++Jt0OSMIHPWoLUpsSX3vq6Ld7De9W4eiovUGEKWNqhrLPUApReOkxLZMtblRK0Gq4LO7KSKdzKUw+u1jjVOKkJWu0QqJ6WsjHtYMW2ePtaFgWlWfuIqE+TElJV5zg2lHqAc4q0b68JgvZVrfMoEmZWyMu5xRKeYqaYcb1rtWBeSuieCArSlrLx1qpC4lLRSzlXxpqJWTpeUCdKcsbb9UUndk7virdI1XsVpTNeatjL+4Xvxjjn3T9q6PK9MmNkpFe9EKa14Rapg7PuZdFQ9r+Iw087lopKKt9T3GoenzJ+Cix6HDwqYdtj7JPw4FbEyyvge3jhA0/9Yw/VDGaYJsKSNZ756AU/s7qXm1qf53GCKltoIh+Kr+OGJ00js6aPmtmf4TH+K6SEhfcNSfrJyDvvcfh2lpnoqpfsDPsNBziRMP18jDsDd/CcG83W2GaqpZIiv8q3DP7ODZn7KrSzkd/wFf5giS3wt3jDA89D4KFy0GWIzIH0GfO4fYNmnQjxd4vFdJ1xB7q/O5L6L57PrwCDR//Iwtzy3i42/38wH5jWy6ZaVPHLbM1z6r3/k0tsv4oG7/8hlbXXs/v4q/nXNbtr+/VWuXTmH/+n26yg14ansJnkWL1DFkzzG9YfPfYF/O/z9j/hzIqM2/X6IT9LCG1Nmg0NJ9VVqz/fuhtlQcQjCCahIQeQk6DUB2Pl+YSu9F89nF8CMWpKNVezf20/jtm4Wf/J01gB88nTWbO1mCcChBO1nt7EJYPlsOgZStGzpeu9OijZSOZXrDsvZTAODRa/lgP2cy7n88fC5P7CEOjppYP8UWgEl1ldZxHsB9HwMHjsLvtMCd1bD0C2wwVgv3ffyWgcth4aYvWoO25NZpi1spRfg1BZ6k1mmAcyoZc8Le1gK8MgW5g6ladneXZ76UDcJmzJ9kL/EqYTpYyEHAOglyno+yp/xUAlGK+lrKrV4BeAtqHkRlvwJ/uEAfDUFkb+D80s8tqfoTBD93hpuvPwU7j2h7r09iyucP7EBuOk8HklmqP70A3zz95u5sKGK3aEKK4M572FKPe/ReINlzCnwug9wJWewmoaSNAUoqXhLfc9rAH4Ai1qhczEMAKyCV1+Fkz8Oz5V4fE8wlCb09dXceMZ0Xrr+bF4FiIboe6uThoWt9L7VSUM0RD/ACXUM/9Ol/AQgZ+Av7+OOM6bT6ab95SBTDr+bpoJ3WMol3Hb4XDfz2M9SXuEastQAhvtJcw1PTcGIJf1IKrV4cwALoOsnML8DItMh9RKctgh2SgDyq3IGvraav2mpYf/XV7B65Py8Jtb9agPLb1nJI7/awPL5TawD2NdPdUOUVG2E7F0vsaKtjs2jPbWNpKUM2XbPs4gaOphDz+FzX+bOw9//H64kzPAUCRdsEO+NsP038PLpcHMF5E6E3XfDsy/buV/Ne3jobU7Z3sP7p0XZ++kH+CbAZafy68+fyyPxp/ncX97HBbURum79ED8AeHU/7T9bz/WAaapm380f5KeuvoAyMaWe939zA90sIEMdt/OPvI8HuZrn2cQy5rF2Ckc6FiW93RFjSqgfkRbgmrEuP3EFyUS9tW1QlAnwk1YS97WUvhKnzLxkYmZdqX55qQNWiaNdjNiZVaNMgqSd6TpDx37K5CntW2bMEEeZOoST9k+blfGxx85su5LGKsrxhhVfLAeiNnbsVSbF7oiVZYH2ijdifQxVGQ9pyHWGVbwTxV3xqudVgK5KK8sBwWbxRtXzKkBHxMpeZjkTMyXdwrQc4h1zQ4uaASuDFMoE2RuxctWhpJFmKI94x0ztq+/VhusK7IlYWV3WU+oByiXeop+skRShUNrKKZMyAXbZuUzUU+oBSv+mGZMBusa6XDNobbBCGSe7o1ZGmrtLPUC5PvEOjHWhtl89b5BJCNmuSivF21PqAVwXb12fZlkFmTdqrNxcG4Ig3voeK+93lHHygp0NflImZo6a1z8VlEc4xvRA8U/Y5oMacQ4qWTBr6qysKiv5/S6Uz/MCThO20VQPURkZJl1GOxSPsDNKMhGycp+id8oxSDnFu22sCw1dKt4gsrbO2mBlRzkGKad490BxkbaMeUes2Mxz9ZR0Fz0XmeoWskUpn3iNyQI7i11q7bBy6qQchZ4Q6Z1RK+MdXSZmyhJBL3ekt+jUuaGbSEXWyvxWZQxeqbX2VqksXhfKL97dFJk6C8i0HmvX+5QiPFdvZT4zWCteZ+pcNOrctluTNYJCd4j0n2pLuwmXi5QlWAXl97wAG4qdnLXD2uCFMorfNpG2dJ+qznIkZ4xQfvEasx84NPp09RCVNf06dbadhJB9qMlar7ujnIO5lZr4ZrGTbXu0wsh2VjcwnKywNiV2RzkHc+tN3EyRns6zdli5dKDkSUPuly3Wet1eEzNjlr6WAnfE6wSujtjIuKGHiKZK2ssL9Qz3VVq7pr+13AO6OX15EziiQdfMXUeeU/xPDszPW62eWQVIvMakgddHn573lkadbeT1Gob3R6wVb5eJmbJUEhXiduBgHaO6S9YOEG44ZP+WlkHj5y2u/6+Vko1uDOruG+r0t3px9Om5mzVV0iZerSGxscbKul1wMgY3uzGw+5+GxmwD9hWemrWDau0qaQcpIfdP7VbfCm0udXP1sXBfvA4vULCLeIVB2nfr1NkG7m1myNIGcyMUzVkoB94QrzFdjEqbPHmjtcGNwNARJnmvfRtmF7LPjUDVCN4Qr8OfKNiYqb6PSEtH6beMUEpDFsx328HSHOYRXPO64CXxGpMEnio8tWidh+xTJsT/ayTxVrW1QSpwVkmKNpcoF94ShzG7gPUjDxu7iDYd1Htfv9ERJvkfM6yeLgO8YmLG1VURb4nXYS1wcOTBaetctESZMFkwd8xEsnZPl/uAt902wnviNSYHPE4+dbLlIFXTutT7+oWftTK0vcrqpSGAl932uuBF8QIY0wc8M/JwkXpfX/CHaQzeb3d0GZyG6lvcNgK8Kl4YSd7YADC9g6qmgxp59jLrahj6X23WCxccr+uJlk3eFa/D8+QjeotfIkRO+1x5kd0Rkt+aRdTyZSFw2tyMuXlAufG2eI0xwGqgo66fyOztRxbwK+7SHSL99dlUWtwdo5Aj8vDdxPtvuFO4/whw6PRXqapMaascr5AQst+YjbG4wL6Qt03M7Dv208qH98ULYEwKeDicZnDh69ppwwtkIHfbiaT3Rq2PLIOT+ecprwt+ES+AMQng93PfxtT2aZdJN8mBuauN4ddrrO1HNZqXTMx4brnSP+IFMKZP4KElaxjU4JU7pCH3nZkMP9kQiMgyQIeJmbfcNqIY/hIvgDE9TV08MmtnefZAVd5loILM104ivaaeardtKRM54Fm3jRgL/4kXwJiB2dt4NA26OWiZeKeS1E1zYbPdxQajecXNkr9j4U/xAq3vmOTfz+HJ12rod9sW29lUxdAX51HZGba6qH40HcCrbhtxNMR4I1lk0khczrj+AEs+0U2t27bYyNP1JL7XTnUAEjAKSQH3mZgZOOYzXcT34gWQuFz64V5ab+qgKuzj2YSXyIH5RQuJ/9sayA/F1V7KpBoLW/7Rn3iygfTn55HdHNUKpOOlN0T6jpkMB1S4b/pBuGCJePPd+x57JwJfnkvVD2eQSIp2n5woOTCPTyNxw3xCLwUnolzIQWCN20aMFyumzSNIXOYClwC0psl8ZT/pM4YC+U84Yd6pJPW9dozF/ZWPRQL4jdfvcwuxSrwAEpdzgHNGHn+kh8QNB4jWmEDk306YDOR+28Twz6ZTbXn3i6ORAR40MdPptiETwTrxAkhcLgHmjjyeliH79/tJLk0EJitoXGyLMvzddip2ByM/eSwM8KiJmV1uGzJRbBVvGLgCmF54ftkAQ5/uROYnA5OTW5SBCjL3tJL6XZN+mAHPm5hxtYXrZLFSvAASlyhwJdA8+tqiBMnrOsmdGbD74Z4Q6QeaST/USFU6GPW3x2K9iRnPVQuNF2vFCyBxqcERcEOx67OTpK7rJHPeANUh7L3f66wkdW8zmccaA31fO5ptJmZWu23E8WC1eAEkLnXAVUDdWM9pTZO59hDJVX1UR4wdHikH5s1qhu9vRl6uC/ZtQhG2A497oQPk8WC9eAEkLg04Aj7qNLkmS/bqboaXD1A5J0mkwofeuDdEek0dqXtbiBwM635PRdgKPOl34UJAxAsgcWnGCWKNywvVZcle0E/ygn44fYho1KNLTVkwO6Ik19aRe76eyp3Bjhwfiy04wrXinz4w4gWQuEwDLgemTeTnQgazZJDkB/vJLR0k3JR116P1V5BZX0PqhXpkbR3RYQ0+jYe3gadtES4ETLwAEpdq4DKgdbK/Y06S1FkJMicPw9wkFe0pwqVIAsmB6Q+R6awk2xEmtzOKWVtHeKv9OxJMNZuAZ20SLgRQvHB4HfgSYNZU/c7GDJnZKTLtKczMFKYtDTPSVEQNSP7euSL/Vo/cSOfdpYiB7kqy+yKYPRHYE0H2RAh1hKnUJZ3j5o8mZjxdlztZAileAIlLBbAKOMVlU5TSkAGe8kuF0GQIrHhHkLicByxx2w5lSkkAj/gtV3miBF68ABKX+cCHQJdWLKATR7jW766h4s0jcWkCPgI0umyKMnm2AM+YmAnErhoq3gLygawVwKlu26JMiAxOgYEn+yuXChVvESQuC4ELIFDdEv3KIZxUxx63DSk3Kt4xyCd0rARmum2LUhQDrAP+ZEOq42RQ8R4DictpwPtBEyM8RB9OmmOgd81Q8RZBRHYANxjjlIzlSwtXUNCdQ3GFDE4j9PUmZgLfYFDFW4TR4j183llSuoBjVCcpJWE7sMZPDeJKjQZkJoCJmW0Sl93AYuBMdF24HPTiRJL3uG2I11Dxjs0yEbkLaAd+A3weeBS4yxhzv0yTevp5gpXcxYW8ybOcxho+yVf5FpuZzkNcxwAnIhha2MCnuIcWhtx8QT5jGCcg9XpQA1LHQsU7Nn8NfBQYBH4H3AI8jZMPfT/9LAe2sYYcF7KNHVxFC28DThz0HB5mGZvppYp7uJHfciWf5V5XXom/GAbW4+xckHbbGC+jFStj8y/GmN3GmC7gduBaHPF+KH99JfBt0iw3MbOaHTTRxLOAYQEHWclGqsnQxgBnsppDLHDpdfiFIeBF4B4TM6+pcI+Net6x2V3w/U6c9d41wAIROQGnmOEqIC4ircBS1nM1nyDJJlbwOLfSwwKyVAFCJdbn2k6SQRxPuzEoaY1ThYp3bGYXfH8SsM8YkxCRl4H/CrxhjEmJyAvAl4GtxjhVLCLy5wjb+Qu+xGmcxB9YyVquLf9L8CwG58NxI7DLtiL5cqHiHZsviMhDOOVlNwO/zJ9/GvgicGf+8VPAt4GfFfxsPYZefsFLVLIT+ArOGmWOYN+qDABvAZtMzAy6bYzfUfGOzT3AYzjT5d8Ct+XPPw18A3im4HF9wWOAOPBToJcMW3CE/aX81/k4DQDa8GF3ykmQxrnt2ALsVi87dWiShkvks7bm5Y827PLIQzjT4u3AHs2GKg0qXg8gcanEEfCs/NGCv7xyBqcIfh/OPewBl+0JBCpeD5LfZ2lm/mjB2W/JS4URA8A7BcchTaQoPypen5DftmVEyCNHPaWLWxigH6eCp7fg6yENNnkDFa/PyXf/qClyRIFQwSG8OxU3QGrUkSz42g/0qzf1NipeRfEpNkU4FSVQqHgVxaeoeBXFp6h4FcWnqHgVxaeoeBXFp6h4FcWnqHiVcSEiO0TkYpfG/lsRec6Nsb2MildRfIqKV5lyRETrxMuAijeAiMjXRGSviPSLyFsicpGI/FhEbit4zioRGd0reZmIbBCRbhH5kYhUFT43/3s7gB+JSIWIfF1EtorIIRG5V0SaC37/r0SkQ0R6ReQZETmj4FqLiDwoIn0ishY4ucRviS9R8QYMEVmI08ZnmTGmHqe97Y5x/vhIO9yTgQU47XBHaMOpdJoDfA64Cfg4TrfNmUA3cHfB8x/G2Up1BvAK8POCa3fjtIBtBz6bP5TRGGP0CNCB04LnAHAxEC44/2PgtoLHq4A9BY93ADcWPL4cp+neyHNTQFXB9Y3ARQWP23Fa4lQWsakRp9KpAacCKg2cVnD9DuA5t987rx3qeQOGMWYL8N+AW4EDIvILERnvNqbF2uGOcNAYM1zweA7waxHpEZEeHDFngRNEJCQi38lPqft41/O3AtNxapRHj6WMQsUbQIwx9xhjVuAIzAD/iNM/uabgaW1FfvSIdriFv3bUc3cDlxljGguOKmPMXuCvgKtxvH8D7+6+KMBBnLY6o8dSRqHiDRgislBELhSRKM595RBOS9rXgMtFpFlE2nC882i+ICIn5gNPhe1wi/F94HYRmZMfd7qIXJ2/Vo9T9H8I5wPjjpEfMsZkgQeAW0WkRkROBz4z6RdsMSre4BEFvoPTMK4DJ2D0DZy2tOtwprCPUVyYI+1wtwFbebcdbjH+GXgQeExE+nG2Mjk/f+2nOFPhvcCG/LVCvgjU5e37MfCj8b+84KCdNBTFp6jnVRSfouJVFJ+i4lUUn6LiVRSfouJVFJ+i4lUUn6LiVRSfouJVFJ+i4lUUn/L/AaZD81vvwMG0AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a=list(v1.mut.unique())\n",
    "b=list(v2.mut.unique())\n",
    "venn2([set(a),set(b)],('bwa','subread'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## contaminants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out temp_blast.txt -outfmt \"6 qseqid sseqid qseq sseq pident qcovs length mismatch gapopen qstart qend sstart send evalue bitscore stitle\" -query temp.fa -db contaminants.fa -evalue 0.001 -max_target_seqs 50 -num_threads 8 -task blastn\n",
      "LT708304.1 Mycobacterium bovis AF2122/97 genome assembly, chromosome: Mycobacterium_bovis_AF2122/97    932\n",
      "NZ_CP021380.2 Pseudomonas aeruginosa strain CCBH4851 chromosome, complete genome                       240\n",
      "Name: stitle, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "reload(tools)\n",
    "\n",
    "def check_unmapped(fastafile):\n",
    "    \"\"\"Check possible source of unmapped reads\"\"\"\n",
    "    \n",
    "    recs=list(SeqIO.parse(fastafile,'fasta'))\n",
    "    recs = recs[:900]\n",
    "    SeqIO.write(recs,'temp.fa','fasta')\n",
    "    tools.make_blast_database('contaminants.fa')\n",
    "    bl = tools.blast_fasta('contaminants.fa', 'temp.fa',threads=8,show_cmd=True)    \n",
    "    print (bl.stitle.value_counts())\n",
    "    return bl\n",
    "\n",
    "bl=check_unmapped('/storage/btbgenie/dup_results/unmapped/TB19-002014.bam.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GCACAGATGGAACGGCTCGCGGTCCCCGACGA                                                                                                                           35\n",
       "TCGTCGGGGACCGCGAGCCGTTCCATCTGTGC                                                                                                                           25\n",
       "TTGAAGGTGTGGAAGAACTCCTCTTGAATACCCTCTTTGCCTTCAATGAATTGGATGTCGTCGACCAACAGCACGTCTACGTCGCGGTAGCTGCGTTTGAATGCG                                                  22\n",
       "ACCAGC---CTTAACCGTCGCTACACCTTTGATACGTTCGTTATCGGCGCCTCCAACCGGTTCGCGCACGCCGCCGCCT                                                                            14\n",
       "GTGTTGGCGTCGGCGATCAGATCGCGAAGCACAATCTCGGCCAGCGCTTTGTCGATTGGTGTTTTGTTCAATGAGGCGAACGCGGTGACCCGGATCAGCGCGCCCTCGAGTTCACGGATATTGCGTTCGATACTGCTGGCGATGAGTTCGA    11\n",
       "                                                                                                                                                           ..\n",
       "GATGAACAGGGGGTTGTAAGCGCGGGCGGGTGCTTCTGCGATCGCCAAGGCGGCGGCGTGCGCGAACCGGTTGGAGGCGCCGATAACGAACGTATCAAAGGTGTAGCGACGGTTAATGCTGGTTACGCCAGCGGTAGCGGAATCGGTATTG     1\n",
       "GTTAAGGCTGGTTACGCCAGCGGTAGCGGAATCGGTATTGCGCGGGCGCTCGGTGAAGTAACTTGGCCAACTGTGCTGGTTATCGCCCCGTGCCGCAGCGCTGTCATCAATCTCGTCGTTGTCGGTTGTGGTGTCTGGCGATGTGGTAGCA     1\n",
       "GGCCGAGATTGTGCTTCGCGATCTGATCGCCGACGCCAACACCATGCAAATCAGCGCGGCGACGATCATGGCTGCCACCGCCGAATACTTCGACACTACCGTCGAAGAGCTTCGCGGGCCCGGCAAGACCCGAGCACTGGCCCAGTCACGA     1\n",
       "CGACGTATCTCCCTGGTTCTCGTTAGT                                                                                                                                 1\n",
       "CCTTCAACACCTTGCACAATGCCAACAAGCAAATCGTCATCTCATCTGACCGCCCACCCAAGCAGCTCGCCGCCCTCGAGGACCGGCTGAGAACCCGCTTTGAGT                                                   1\n",
       "Name: qseq, Length: 706, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bl.qseq.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## find all multimapping regions in bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b''"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf = '/storage/btbgenie/dup_results/mapped/TB19-003437.bam'\n",
    "out = 'multimapped.bam'\n",
    "threads=4\n",
    "cmd = 'samtools view -@ {t} -b -f 256 -F 4 -h  {b} > {o}'.format(b=bf,o=out,t=threads)\n",
    "subprocess.check_output(cmd,shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bamnostic as bs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_multimapped():\n",
    "    bam = bs.AlignmentFile('/storage/btbgenie/dup_results/mapped/TB19-003437.bam', 'rb')\n",
    "    i=0\n",
    "    rc=[]\n",
    "\n",
    "    for read in bam:\n",
    "        if i > 1000000:\n",
    "            break\n",
    "        #if read.mapq == 0:\n",
    "        rc.append(read.pos)\n",
    "        i+=1\n",
    "    \n",
    "get_multimapped()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 50s, sys: 47.8 ms, total: 1min 50s\n",
      "Wall time: 1min 50s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "get_multimapped()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "608012     13\n",
       "1342643    13\n",
       "892029     12\n",
       "1170985    11\n",
       "608386     11\n",
       "           ..\n",
       "1343378     1\n",
       "104339      1\n",
       "104343      1\n",
       "608153      1\n",
       "401408      1\n",
       "Length: 4109, dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x=pd.Series(rc)\n",
    "x.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam = bs.AlignmentFile('multimapped.bam', 'rb')\n",
    "res=[]\n",
    "for read in bam:\n",
    "    res.append((read.pos,read.seq,read.mapq))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame(res,columns=['pos','seq','qual'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          32\n",
       "1306895    17\n",
       "2767810     8\n",
       "3080741     6\n",
       "1306896     5\n",
       "4317352     5\n",
       "3351441     4\n",
       "1306875     4\n",
       "1306879     3\n",
       "956926      3\n",
       "Name: pos, dtype: int64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAEFCAYAAAD69rxNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAASZElEQVR4nO3dbYxcV33H8d9vxxOyJqh2lEnqbAxuo+BC08YuKydqJMRTahNaYopoa0GUF1HNC6iCRF3FBIlQFYHkAn1DkUwT4YqQNijGRIBwrTSIBwXDGjtxHGMlQAheW/GSZEtCFrJe//ti7mz2YWbnee492e9HWs3MmfvwnzMzP1/de+bYESEAQHqG8i4AANAZAhwAEkWAA0CiCHAASBQBDgCJWjHInV100UWxbt26Qe4SAJJ36NChX0VEZWH7QAN83bp1GhsbG+QuASB5tn9Rr51TKACQKAIcABJFgANAoghwAEgUAQ4AiRroKJRO7Ds8rl37T+jU5JQuXTWsHZvXa+vGkbzLAoDcFTrA9x0e1869RzU1PSNJGp+c0s69RyWJEAew7BX6FMqu/Sdmw7tmanpGu/afyKkiACiOQgf4qcmpttoBYDkpdIBfumq4rXYAWE4KHeA7Nq/XcLk0r224XNKOzetzqggAiqPQFzFrFyoZhQIAixU6wKVqiBPYALBYoU+hAAAaI8ABIFEEOAAkigAHgEQR4ACQqKYBbvt82z+0/ZDtY7Y/nrVfaPuA7cey29X9LxcAUNPKEfjvJL0lIq6StEHSFtvXSLpV0v0RcYWk+7PHAIABaRrgUfV89rCc/YWkGyTtydr3SNrajwIBAPW1dA7cdsn2EUlnJB2IiIOSLomI05KU3V7cYN3ttsdsj01MTPSobABASwEeETMRsUHSZZI22b6y1R1ExO6IGI2I0Uql0mGZAICF2hqFEhGTkr4taYukp2yvkaTs9kyviwMANNbKKJSK7VXZ/WFJb5P0E0n3SbopW+wmSV/rU40AgDpamcxqjaQ9tkuqBv49EfF12w9Kusf2zZKelPSePtYJAFigaYBHxMOSNtZpf1rSW/tRFACgOX6JCQCJIsABIFEEOAAkigAHgEQR4ACQKAIcABJFgANAoghwAEgUAQ4AiSLAASBRBDgAJIoAB4BEEeAAkCgCHAASRYADQKIIcABIFAEOAIkiwAEgUQQ4ACSKAAeARBHgAJAoAhwAEkWAA0Cimga47bW2H7B93PYx27dk7bfbHrd9JPu7vv/lAgBqVrSwzFlJH46IH9t+laRDtg9kz302Iv61f+UBABppGuARcVrS6ez+c7aPSxrpd2EAgKW1dQ7c9jpJGyUdzJo+aPth23faXt1gne22x2yPTUxMdFctAGBWywFu+wJJ90r6UET8WtLnJV0uaYOqR+ifrrdeROyOiNGIGK1UKt1XDACQ1GKA2y6rGt53RcReSYqIpyJiJiLOSfqCpE39KxMAsFAro1As6Q5JxyPiM3Pa18xZ7F2SHul9eQCARloZhXKtpBslHbV9JGv7iKRttjdICklPSHp/H+oDADTQyiiU70lynae+2ftyAACt4peYAJAoAhwAEkWAA0CiCHAASBQBDgCJIsABIFEEOAAkigAHgEQR4ACQKAIcABJFgANAoghwAEgUAQ4AiSLAASBRBDgAJIoAB4BEEeAAkCgCHAASRYADQKIIcABIFAEOAIkiwAEgUQQ4ACSqaYDbXmv7AdvHbR+zfUvWfqHtA7Yfy25X979cAEBNK0fgZyV9OCJeJ+kaSR+w/XpJt0q6PyKukHR/9hgAMCBNAzwiTkfEj7P7z0k6LmlE0g2S9mSL7ZG0tU81AgDqaOscuO11kjZKOijpkog4LVVDXtLFDdbZbnvM9tjExESX5QIAaloOcNsXSLpX0oci4tetrhcRuyNiNCJGK5VKJzUCAOpoKcBtl1UN77siYm/W/JTtNdnzaySd6U+JAIB6WhmFYkl3SDoeEZ+Z89R9km7K7t8k6Wu9Lw8A0MiKFpa5VtKNko7aPpK1fUTSpyTdY/tmSU9Kek9fKgQA1NU0wCPie5Lc4Om39rYcAECr+CUmACSKAAeARBHgAJAoAhwAEkWAA0CiCHAASBQBDgCJIsABIFEEOAAkigAHgES1MhcKlpF9h8e1a/8JnZqc0qWrhrVj83pt3TiSd1kA6iDAMWvf4XHt3HtUU9MzkqTxySnt3HtUkghxoIA4hYJZu/afmA3vmqnpGe3afyKnigAshQDHrFOTU221A8gXAY5Zl64abqsdQL4IcMzasXm9hsuleW3D5ZJ2bF6fU0UAlsJFTMyqXahkFAqQBgIc82zdOEJgA4ngFAoAJIoAB4BEEeAAkCgCHAAS1TTAbd9p+4ztR+a03W573PaR7O/6/pYJAFiolSPwL0raUqf9sxGxIfv7Zm/LAgA00zTAI+I7kp4ZQC0AgDZ0cw78g7Yfzk6xrO5ZRQCAlnQa4J+XdLmkDZJOS/p0owVtb7c9ZntsYmKiw90BABbqKMAj4qmImImIc5K+IGnTEsvujojRiBitVCqd1gkAWKCjALe9Zs7Dd0l6pNGyAID+aDoXiu27Jb1J0kW2T0r6mKQ32d4gKSQ9Ien9/SsRAFBP0wCPiG11mu/oQy0AgDbwS0wASBQBDgCJIsABIFEEOAAkigAHgEQR4ACQKAIcABJFgANAoghwAEgUAQ4AiSLAASBRBDgAJIoAB4BEEeAAkCgCHAASRYADQKIIcABIFAEOAIkiwAEgUQQ4ACSKAAeARBHgAJAoAhwAEtU0wG3fafuM7UfmtF1o+4Dtx7Lb1f0tEwCwUCtH4F+UtGVB262S7o+IKyTdnz0GAAxQ0wCPiO9IemZB8w2S9mT390ja2tuyAADNdHoO/JKIOC1J2e3FvSsJANCKvl/EtL3d9pjtsYmJiX7vDgCWjU4D/CnbayQpuz3TaMGI2B0RoxExWqlUOtwdAGChTgP8Pkk3ZfdvkvS13pQDAGhVK8MI75b0oKT1tk/avlnSpyRdZ/sxSddljwEAA7Si2QIRsa3BU2/tcS0AgDbwS0wASBQBDgCJIsABIFEEOAAkqulFTAD523d4XLv2n9CpySldumpYOzav19aNI3mXhZwR4EDB7Ts8rp17j2pqekaSND45pZ17j0oSIb7McQoFKLhd+0/MhnfN1PSMdu0/kVNFKAoCHCi4U5NTbbVj+SDAgYK7dNVwW+1YPghwoOB2bF6v4XJpXttwuaQdm9fnVBGKgouYQMHVLlQyCgULEeBAArZuHCGwsQinUAAgUQQ4ACSKAAeARBHgAJAoAhwAEsUoFGAZY5KstBHgwDLFJFnp4xQKsEwxSVb6CHBgmWKSrPQR4MAyxSRZ6SPAgWWKSbLS19VFTNtPSHpO0oyksxEx2ouikB9GJSwfTJLVP4P6HvViFMqbI+JXPdgOcsaohOWHSbJ6b5DfI06hYBajEoDuDfJ71G2Ah6T/sX3I9vZ6C9jebnvM9tjExESXu0M/MSoB6N4gv0fdBvi1EfFnkt4u6QO237hwgYjYHRGjETFaqVS63B36iVEJQPcG+T3qKsAj4lR2e0bSVyVt6kVRyAejEoDuDfJ71PFFTNuvlDQUEc9l9/9C0j/3rDIMHKMSgO4N8nvUzSiUSyR91XZtO1+OiG/1pCoAyWEI6ksGNbqn4wCPiJ9JuqqHtSBnDCNEp/js5INhhJjFMEJ0is9OPghwzGIYITrFZycfBDhmMYwQneKzkw8CHLMYRohO8dnJB/8jT0L6fZV/68YRjf3iGd198JeaiVDJ1rvf0PrV9KXqY4TCYkXqk25rYQhqPgjwRAziKv++w+O699C4ZiIkSTMRuvfQuEZfc2HTfSxVnyRGKCxQpFEbvaqFibEGj1MoiRjEVf5u9rHUuoxQWKxIfVKkWtAejsATMYir/N3so5N1l/MIhSKN2ihSLWgPR+CJGMRV/m72sdS6jFBYrEh9UqRa0B4CPBGDuMrfzT6WWpcRCosVqU+KVAvawymURAziKn83+2hlXUYovKRIozaKVAva48hGHAzC6OhojI2Ntb3eR/cdnTe0bdvVa/UvW/9EUvUK+u33HdPk1HTdda3q/zoxUmdY29z1Vq8s62N/9ceSqh/k8ckplWzNRGj1yrKe/+20ps81r3XI0rmQbKnWtatXlvWOP12jbzx8Ws++ML/OV55X0m9enKmzpZdq+srYk/r+T59pvvMGykPSBeeX9ewL07OvqdYfknTbV482rGGhleUhhaSpOZ0xJKmFrqnrkledp4O3Xaf3fuHBea+xPKSG/T23byWpZOsPKyv1+JnfKGbbqu9Do8eN1LZd66dVw2XZ0uQL07PBJtXvs9p7L0mrhsu6/Z3137vaNhd+FmrPvXh2Ri/MefFL9cVSzitZ0zNR9zUPl4f02+lzWrWyrAjp/6bmv77ad2De61O1/zpJjFesGNLvzlZfRK1vtm4cafr9rVlZrp4seKFOR9QyYfQ1F2rHV4407KtrL79QTzw9pfHJqXnvVe09r/de12pcmAmdsqX3Xv3q2fxqfT0fqvd/Dhc+wD+676i+9IMnF7W/75pXZ2/YQ5o+19prGC6X9Mm/rnZcvfVKQ9aQ1PL2Ulceqn4Y8365KyydTaTLi9Jn/VIesmRpeqa/L7A8ZP3tprX67x/+srDft+FySe9+w4juPTS+aJROt953TXshnmyAX77zm3X/xSvZ+v3fO3/RUUIzI9mFmXbXA9Bb3R7NDkK/aizZ+uknr295+UYBXvhz4I06byaio2FODI0CiqHo4S31r8Zebbfwo1BK1f8wom57J8OclhrWBmBwGn23i6RfNfZqu4UP8G1Xr23YvmPz+uo5uxbNHdZWb73SkNvaXurKQ1YRXu6KAtTQqqL0Wb+Uh6xyqf8vsDxUvfBY5O/bcLmkbVevXTTEshca5Vq7Cn8O/OpPHNBTz73Yp4oAYPCe+NQ72lq+0TnwQh+BE94AXo7W3fqNnmyn0AFOeANAY4UOcABAYwQ4ACSq0AF+yavOy7sEACisrgLc9hbbJ2w/bvvWXhVVc/C26whxAC877Y5CaaTjX2LaLkn6nKTrJJ2U9CPb90XEoz2pLHPwtut6uTkAeNno5gh8k6THI+JnEfGipP+SdENvygIANNNNgI9I+uWcxyeztnlsb7c9ZntsYmKii90BAObqJsDr/QZ20c86I2J3RIxGxGilUulidwCAuboJ8JOS5v6g/zJJp7orBwDQqm4C/EeSrrD9B7bPk/R3ku7rTVkAgGa6mszK9vWS/k1SSdKdEfGJJstPSPpFh7u7SNKvOlz35Yx+qY9+qY9+qa/o/fKaiFh0DnqgsxF2w/ZYvdm4ljv6pT76pT76pb5U+6XQv8QEADRGgANAolIK8N15F1BQ9Et99Et99Et9SfZLMufAAQDzpXQEDgCYgwAHgEQlEeD9nrY2RbbvtH3G9iN511IkttfafsD2cdvHbN+Sd01FYPt82z+0/VDWLx/Pu6aisF2yfdj21/OupV2FD/A509a+XdLrJW2z/fp8qyqEL0rakncRBXRW0ocj4nWSrpH0AT4vkqTfSXpLRFwlaYOkLbavybekwrhF0vG8i+hE4QNcTFtbV0R8R9IzeddRNBFxOiJ+nN1/TtUv5qJZMpebqHo+e1jO/pb9CAbbl0l6h6T/yLuWTqQQ4C1NWwssZHudpI2SDuZcSiFkpwqOSDoj6UBE0C/VqUD+SdK5nOvoSAoB3tK0tcBcti+QdK+kD0XEr/OupwgiYiYiNqg6c+gm21fmXFKubP+lpDMRcSjvWjqVQoAzbS3aYrusanjfFRF7866naCJiUtK3xTWUayW90/YTqp6afYvtL+VbUntSCHCmrUXLbFvSHZKOR8Rn8q6nKGxXbK/K7g9Lepukn+RaVM4iYmdEXBYR61TNlf+NiPflXFZbCh/gEXFW0gcl7Vf1gtQ9EXEs36ryZ/tuSQ9KWm/7pO2b866pIK6VdKOqR1NHsr/r8y6qANZIesD2w6oeFB2IiOSGzWE+fkoPAIkq/BE4AKA+AhwAEkWAA0CiCHAASBQBDgB90u6kc7b/xvaj2YRjX266PKNQAKA/bL9R0vOS/jMilvzlq+0rJN2j6qRjz9q+OCLOLLUOR+AA0Cf1Jp2zfbntb9k+ZPu7tv8oe+rvJX0uIp7N1l0yvCUCHAAGbbekf4iIN0j6R0n/nrW/VtJrbX/f9g9sN53qYEUfiwQAzJFNsvbnkr5SnfVBkvSK7HaFpCskvUnVOZ++a/vKbO6aughwABicIUmT2ayQC52U9IOImJb0c9snVA30Hy21MQDAAGRTG//c9nuk6uRrtq/Knt4n6c1Z+0WqnlL52VLbI8ABoE8aTDr3Xkk3235I0jG99D+M7Zf0tO1HJT0gaUdEPL3k9hlGCABp4ggcABJFgANAoghwAEgUAQ4AiSLAASBRBDgAJIoAB4BE/T895/fY3ORVjwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "cts=df.pos.value_counts()\n",
    "plt.scatter(x=cts.index,y=cts)\n",
    "cts[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## igv browser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import igv\n",
    "ref = {\n",
    "    \"id\": \"AF212297\",  \n",
    "    \"fastaURL\": \"Mbovis_AF212297.fa\",\n",
    "    \"indexURL\": \"Mbovis_AF212297.fa.fai\",  \n",
    "  }\n",
    "\n",
    "b = igv.Browser({\"reference\": ref})\n",
    "b.load_track(\n",
    "    {\n",
    "        \"name\": \"Local VCF\",\n",
    "        \"url\": \"duplicate_test/snps.vcf.gz\",\n",
    "        \"format\": \"vcf\",\n",
    "        \"type\": \"variant\",\n",
    "        \"indexed\": False\n",
    "    })\n",
    "\n",
    "b.load_track(\n",
    "    {\n",
    "        \"name\": \"TB19-002014\",\n",
    "        \"url\": \"/storage/btbgenie/dup_results/mapped/TB19-002014.bam\",\n",
    "        \"indexURL\": \"/storage/btbgenie/dup_results/mapped/TB19-002014.bam.bai\",\n",
    "        \"format\": \"bam\",\n",
    "        \"type\": \"alignment\"\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b.show()\n",
    "b.zoom_in()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sample</th>\n",
       "      <th>filename</th>\n",
       "      <th>pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR125601_1</td>\n",
       "      <td>ERR125601</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125601_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR125602_2</td>\n",
       "      <td>ERR125602</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125602_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR125614_1</td>\n",
       "      <td>ERR125614</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125614_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR125624_2</td>\n",
       "      <td>ERR125624</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125624_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR125611_2</td>\n",
       "      <td>ERR125611</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125611_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ERR125627_2</td>\n",
       "      <td>ERR125627</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125627_...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>ERR125606_2</td>\n",
       "      <td>ERR125606</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125606_...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ERR125619_1</td>\n",
       "      <td>ERR125619</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125619_...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ERR125603_2</td>\n",
       "      <td>ERR125603</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125603_...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ERR125598_1</td>\n",
       "      <td>ERR125598</td>\n",
       "      <td>/storage/btbgenie/mbovis_ireland/NI/ERR125598_...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           name     sample                                           filename  pair\n",
       "0   ERR125601_1  ERR125601  /storage/btbgenie/mbovis_ireland/NI/ERR125601_...     1\n",
       "1   ERR125602_2  ERR125602  /storage/btbgenie/mbovis_ireland/NI/ERR125602_...     1\n",
       "2   ERR125614_1  ERR125614  /storage/btbgenie/mbovis_ireland/NI/ERR125614_...     1\n",
       "3   ERR125624_2  ERR125624  /storage/btbgenie/mbovis_ireland/NI/ERR125624_...     1\n",
       "4   ERR125611_2  ERR125611  /storage/btbgenie/mbovis_ireland/NI/ERR125611_...     1\n",
       "..          ...        ...                                                ...   ...\n",
       "57  ERR125627_2  ERR125627  /storage/btbgenie/mbovis_ireland/NI/ERR125627_...     2\n",
       "58  ERR125606_2  ERR125606  /storage/btbgenie/mbovis_ireland/NI/ERR125606_...     2\n",
       "59  ERR125619_1  ERR125619  /storage/btbgenie/mbovis_ireland/NI/ERR125619_...     2\n",
       "60  ERR125603_2  ERR125603  /storage/btbgenie/mbovis_ireland/NI/ERR125603_...     2\n",
       "61  ERR125598_1  ERR125598  /storage/btbgenie/mbovis_ireland/NI/ERR125598_...     2\n",
       "\n",
       "[62 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files=app.get_files_from_paths('/storage/btbgenie/mbovis_ireland/NI/')\n",
    "df = app.get_samples(files)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "p=pd.pivot_table(df,index='sample',columns='pair',values=['filename','name'],aggfunc='first')\n",
    "c=list(zip(p.columns.get_level_values(0),p.columns.get_level_values(1)))\n",
    "p.columns = [i[0]+str(i[1]) for i in c]\n",
    "p=p.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
